{
 "cells": [
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-29T15:45:05.606367Z",
     "start_time": "2025-01-29T15:45:02.739386Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "import ViennaRNA as vrna\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "import pyranges as pr\n",
    "import json\n",
    "from collections import Counter\n",
    "from scipy.optimize import curve_fit\n",
    "import scipy.special\n",
    "from scipy.stats import poisson\n",
    "\n",
    "import polygraph.input, polygraph.sequence, polygraph.visualize\n",
    "\n",
    "import src.human_sequences,src.constants,src.rna_analysis,src.rng_sequences"
   ],
   "id": "f50959a437a2866a",
   "outputs": [],
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "id": "d6eae7238000e4b2",
   "metadata": {},
   "source": [
    " # Generating data files"
   ]
  },
  {
   "cell_type": "code",
   "id": "8b92620741b250c5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-29T15:46:35.743712Z",
     "start_time": "2025-01-29T15:46:35.738869Z"
    }
   },
   "source": [
    "'''Defining parameters'''\n",
    "\n",
    "seqnum=100000\n",
    "seqlen=160\n",
    "myu=5.345801185151601\n",
    "sigma=0.8416777786431029\n",
    "factor_a=66009.38523606668\n",
    "params=[myu,sigma,factor_a]\n",
    "np.random.seed(12345)"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "markov_matrix=np.load('data/human_matrix.npy')\n",
    "print(markov_matrix)"
   ],
   "id": "341c0cfe835c5494",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-29T15:45:09.375043Z",
     "start_time": "2025-01-29T15:45:09.371075Z"
    }
   },
   "cell_type": "code",
   "source": [
    "with open('data/mononucleotides.json','r') as f:\n",
    "    mononucleotides=json.load(f)"
   ],
   "id": "55f4c970eef698cc",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "markov_seqs=src.rng_sequences.markov_random_sequence(markov_matrix,params,8000,100,group_name=\"Markov50\",seed=12345)\n",
    "markov_seqs"
   ],
   "id": "880f844cc736e9b8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-29T15:46:44.096523Z",
     "start_time": "2025-01-29T15:46:40.027974Z"
    }
   },
   "cell_type": "code",
   "source": "new_seqs=src.rng_sequences.randomseqs_type2(seqnum,seqlen,group_name=\"monos_adjusted\",seed=12345,probabilities=mononucleotides)",
   "id": "6aeaf43926977fe4",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "seqs1 = src.rng_sequences.randomseqs(seqnum,seqlen,group_name=\"Random_gc_adjusted\",seed=12345,gc_content=0.6081)\n",
    "seqs2 = src.rng_sequences.randomseqs(seqnum,seqlen,group_name=\"Random_gc50\",seed=12345)"
   ],
   "id": "8de06b91ac84f799",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "seqs5 = src.rng_sequences.randomlengthseqs(params,8000,100000,group_name=\"Random_gc_fish\",seed=12345,gc_content=0.4437)\n",
    "seqs5"
   ],
   "id": "fa84b2d355cd02b8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-27T10:52:58.727837Z",
     "start_time": "2024-11-27T10:52:37.857671Z"
    }
   },
   "cell_type": "code",
   "source": "seqs6=src.rng_sequences.randomlengthseqs(params,8000,100000,group_name=\"Random_gc50\",seed=12345)",
   "id": "19205485a58169a6",
   "outputs": [],
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "id": "789af7770c6de70a",
   "metadata": {},
   "source": "np.savetxt('data/small_dataset.txt' , seqs5, delimiter='\\t',newline='\\n',fmt='%s')",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4cbd8b5e0035e7b3",
   "metadata": {},
   "source": [
    "testseqs=polygraph.seqs=polygraph.input.read_seqs('data/small_dataset.txt')\n",
    "testseqs"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "48b1df639ad5802c",
   "metadata": {},
   "source": [
    "seqs3=src.rng_sequences.randomlengthseqs(params,8000,seqnum,\"Random_dl_gc50\",seed=12345)\n",
    "seqs4=src.rng_sequences.randomlengthseqs(params,8000,seqnum,\"Random_dl_gc_adjusted\",seed=12345,gc_content=0.6081)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8683b3308497543c",
   "metadata": {},
   "source": [
    "'''This code is used to generate a dataset with continous IDs and save it in a txt file'''\n",
    "randomseqs=np.concatenate([seqs5,seqs6],axis=0)\n",
    "randomseqsindex=np.array([f\"seq_{i+1}\" for i in range(len(randomseqs))])\n",
    "randomseqs=np.vstack((randomseqsindex,randomseqs.T)).T\n",
    "np.savetxt(\"data/random_dataset.txt\",randomseqs,delimiter=\"\\t\",newline=\"\\n\",fmt='%s')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "c81cd8f6511e1740",
   "metadata": {},
   "source": [
    "'''Here we read out the human genome sequences and save it in a txt file'''\n",
    "humanseqs=src.human_sequences.readFASTA(\"data/gencode_5utrs_human.fa\")\n",
    "humanseqs=src.human_sequences.delDupes(humanseqs)\n",
    "np.savetxt(\"data/humanseqsRNA.txt\",humanseqs,delimiter=\"\\t\",newline=\"\\n\",fmt='%s')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e06a688d3ebe8993",
   "metadata": {},
   "source": [
    "''' Here we write every Sequence into one txt file'''\n",
    "seqfiles=['data/humanseqsRNA.txt','data/random_dataset.txt']\n",
    "with open('data/full_dataset.txt','w') as outfile:\n",
    "    for fname in seqfiles:\n",
    "        with open(fname) as infile:\n",
    "            for line in infile:\n",
    "                outfile.write(line)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5c0239d53acf1e36",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-10-28T14:27:46.537743Z",
     "start_time": "2024-10-28T14:27:46.122663Z"
    }
   },
   "source": [
    "seqs=polygraph.input.read_seqs('data/small_dataset.txt')\n",
    "seqs"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                    Sequence           Group\n",
       "SeqID                                                                       \n",
       "seq_0      UCAGGAUCCUGACGUGAAAUGUUCAGCAAUGCACUGUUUAGCAUGC...  Random_gc_fish\n",
       "seq_1      ACCCGUCUGACGGGUUAAAGUGCUUACAUGAUGUAACUAGGGGAAA...  Random_gc_fish\n",
       "seq_2      GAUUGAGCAGUACGCAUGCAGAGGCGGGUGAUUUUUGAGCGGCAGU...  Random_gc_fish\n",
       "seq_3      UAAUGUCCACCAAAUAAUAUGGGGAUGGAAUACCUGUACGAAUGCA...  Random_gc_fish\n",
       "seq_4      CUCGAAAGGGUCAACUCUAACAAGGGCCGGCUUAGUGUUUUAAGUG...  Random_gc_fish\n",
       "...                                                      ...             ...\n",
       "seq_99995  UAGAAACGGAAGAAUCUAAUCCGAGCCGACACUCAGUAGAUAGCGG...  Random_gc_fish\n",
       "seq_99996  CAAGUCAUGGAGGUGAUAAAGAUGCGGUGGAGAACAGAUCGUUCGG...  Random_gc_fish\n",
       "seq_99997  CUGAGUGGUUUCAGUGCCGUAACAAAUAUACACAUAGUUACGGGGA...  Random_gc_fish\n",
       "seq_99998  UAGUAAGAUACAAAUUGCACUGAGGCAUAUAUGGAAAUACAGAAUU...  Random_gc_fish\n",
       "seq_99999  UCGCGCUACUUAGCUCAUAAAGUCAGUCGGCGUGUGCUCUUCGGGG...  Random_gc_fish\n",
       "\n",
       "[100000 rows x 2 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SeqID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>seq_0</th>\n",
       "      <td>UCAGGAUCCUGACGUGAAAUGUUCAGCAAUGCACUGUUUAGCAUGC...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_1</th>\n",
       "      <td>ACCCGUCUGACGGGUUAAAGUGCUUACAUGAUGUAACUAGGGGAAA...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_2</th>\n",
       "      <td>GAUUGAGCAGUACGCAUGCAGAGGCGGGUGAUUUUUGAGCGGCAGU...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_3</th>\n",
       "      <td>UAAUGUCCACCAAAUAAUAUGGGGAUGGAAUACCUGUACGAAUGCA...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_4</th>\n",
       "      <td>CUCGAAAGGGUCAACUCUAACAAGGGCCGGCUUAGUGUUUUAAGUG...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_99995</th>\n",
       "      <td>UAGAAACGGAAGAAUCUAAUCCGAGCCGACACUCAGUAGAUAGCGG...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_99996</th>\n",
       "      <td>CAAGUCAUGGAGGUGAUAAAGAUGCGGUGGAGAACAGAUCGUUCGG...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_99997</th>\n",
       "      <td>CUGAGUGGUUUCAGUGCCGUAACAAAUAUACACAUAGUUACGGGGA...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_99998</th>\n",
       "      <td>UAGUAAGAUACAAAUUGCACUGAGGCAUAUAUGGAAAUACAGAAUU...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_99999</th>\n",
       "      <td>UCGCGCUACUUAGCUCAUAAAGUCAGUCGGCGUGUGCUCUUCGGGG...</td>\n",
       "      <td>Random_gc_fish</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100000 rows × 2 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "mfe_seqs=pd.read_csv('data/MFE_dl_stats.csv')\n",
    "mfe_seqs.set_index('SeqID',inplace=True)\n",
    "mfe_seqs['Sequence Length']=mfe_seqs.Sequence.apply(len)\n",
    "mfe_seqs"
   ],
   "id": "845b3919b86a16b9",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "63e1273ddfa9072a",
   "metadata": {},
   "source": [
    "# Generating statistics for the Sequences "
   ]
  },
  {
   "cell_type": "code",
   "id": "ed57301afa94683d",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-10-28T14:28:02.135501Z",
     "start_time": "2024-10-28T14:28:02.101978Z"
    }
   },
   "source": [
    "'''Sequence lengths'''\n",
    "seqs['Sequence Length']=seqs.Sequence.apply(len)"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "id": "495ffbd1ee65595a",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-10-28T13:26:07.526136Z",
     "start_time": "2024-10-28T13:26:07.266723Z"
    }
   },
   "source": [
    "'''GC Content'''\n",
    "seqs['GC Content']=polygraph.sequence.gc(seqs)"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "''' MFE '''\n",
    "secstructs=[]\n",
    "mfes=[]\n",
    "for seq in seqs[\"Sequence\"].iloc[0:1000]:\n",
    "    (ss,mfe)=vrna.fold(seq)\n",
    "    secstructs.append(ss)\n",
    "    mfes.append(mfe)\n",
    "#seqs['MFE']=mfes\n",
    "#seqs['ss']=secstructs\n"
   ],
   "id": "ba6b69bb818e133d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-10-28T15:00:39.501089Z",
     "start_time": "2024-10-28T15:00:37.596155Z"
    }
   },
   "cell_type": "code",
   "source": "seqs.to_csv('data/small_dataset.csv')",
   "id": "81fe5a41ef7c3fad",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-10-28T14:58:45.879201Z",
     "start_time": "2024-10-28T14:34:43.188809Z"
    }
   },
   "cell_type": "code",
   "source": [
    "'''sliding window mfe'''\n",
    "seqs['sliding window mfe']=src.rna_analysis.sliding_window_mfe(seqs)"
   ],
   "id": "658d44df40bc31e5",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Progress: |████████████████----------------------------------| 33.5% Complete\r"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m                         Traceback (most recent call last)",
      "\u001B[0;32m/var/tmp/ipykernel_1857496/4212556271.py\u001B[0m in \u001B[0;36m?\u001B[0;34m()\u001B[0m\n\u001B[0;32m----> 2\u001B[0;31m \u001B[0;34m'''sliding window mfe'''\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m\u001B[1;32m      3\u001B[0m \u001B[0mseqs\u001B[0m\u001B[0;34m[\u001B[0m\u001B[0;34m'sliding window mfe'\u001B[0m\u001B[0;34m]\u001B[0m\u001B[0;34m=\u001B[0m\u001B[0msrc\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mrna_analysis\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0msliding_window_mfe\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mseqs\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n",
      "\u001B[0;32m/fast/AG_Ohler/jdemoli/bachelorgit/src/rna_analysis.py\u001B[0m in \u001B[0;36m?\u001B[0;34m(seqs)\u001B[0m\n\u001B[1;32m    410\u001B[0m         \u001B[0mlowest_mfe\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mfloat\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m'inf'\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    411\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    412\u001B[0m         \u001B[0;32mfor\u001B[0m \u001B[0mwin\u001B[0m \u001B[0;32min\u001B[0m \u001B[0mwins\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    413\u001B[0m             \u001B[0mstring\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mstr\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mwin\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0;32m--> 414\u001B[0;31m             \u001B[0;34m(\u001B[0m\u001B[0mss\u001B[0m\u001B[0;34m,\u001B[0m \u001B[0mmfe\u001B[0m\u001B[0;34m)\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mvrna\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mfold\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mstring\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m\u001B[1;32m    415\u001B[0m             \u001B[0mmfe\u001B[0m \u001B[0;34m=\u001B[0m \u001B[0mmfe\u001B[0m \u001B[0;34m/\u001B[0m \u001B[0mlen\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0mstring\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    416\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m    417\u001B[0m             \u001B[0;32mif\u001B[0m \u001B[0mmfe\u001B[0m \u001B[0;34m<\u001B[0m \u001B[0mlowest_mfe\u001B[0m\u001B[0;34m:\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n",
      "\u001B[0;32m~/miniconda3/envs/testenv/lib/python3.9/site-packages/RNA/RNA.py\u001B[0m in \u001B[0;36m?\u001B[0;34m(*args)\u001B[0m\n\u001B[1;32m   8166\u001B[0m     \u001B[0;32mand\u001B[0m \u001B[0mthe\u001B[0m \u001B[0mdata\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m   8167\u001B[0m     \u001B[0mstructure\u001B[0m \u001B[0mRNA\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mfold_compound\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m)\u001B[0m \u001B[0minstead\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m   8168\u001B[0m \u001B[0;34m\u001B[0m\u001B[0m\n\u001B[1;32m   8169\u001B[0m     \"\"\"\n\u001B[0;32m-> 8170\u001B[0;31m     \u001B[0;32mreturn\u001B[0m \u001B[0m_RNA\u001B[0m\u001B[0;34m.\u001B[0m\u001B[0mfold\u001B[0m\u001B[0;34m(\u001B[0m\u001B[0;34m*\u001B[0m\u001B[0margs\u001B[0m\u001B[0;34m)\u001B[0m\u001B[0;34m\u001B[0m\u001B[0;34m\u001B[0m\u001B[0m\n\u001B[0m",
      "\u001B[0;31mKeyboardInterrupt\u001B[0m: "
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "cell_type": "code",
   "id": "8162b12eee8b0c8a",
   "metadata": {},
   "source": [
    "'''uORF counts'''\n",
    "AUG_counts1=src.rna_analysis.uORFs(seqs,'AUG')\n",
    "CUG_counts1=src.rna_analysis.uORFs(seqs,'CUG')\n",
    "GUG_counts1=src.rna_analysis.uORFs(seqs,'GUG')\n",
    "UUG_counts1=src.rna_analysis.uORFs(seqs,'UUG')\n",
    "ACG_counts1=src.rna_analysis.uORFs(seqs,'ACG')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "counts1=src.rna_analysis.counts_concat_all(seqs,AUG_counts1,CUG_counts1,GUG_counts1,UUG_counts1,ACG_counts1)\n",
    "counts1.to_csv('data/uORF_stats.csv')"
   ],
   "id": "7690df01df4a1c7d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "771b1edc4bc44108",
   "metadata": {},
   "source": "seqs=pd.concat([seqs,counts],axis=1)\n",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "seqs=pd.read_csv('data/uORF_stats.csv')\n",
    "seqs.set_index('SeqID',inplace=True)\n",
    "seqs"
   ],
   "id": "1921362bbfe892fe",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "seqs.to_csv('data/uORF_stats.csv')",
   "id": "9d62ae7ef552f06c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "humanseqs=seqs.loc[seqs['Group']=='Human']",
   "id": "6a787ded6dfb9963",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "cleanedhumanseqs=src.human_sequences.readFASTA(\"data/gencode_5utrs_human_cleaned.fa\")",
   "id": "35416bcc1218257f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "cleanedhumanseqs",
   "id": "35f873cdf324e8c6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "seqs = []\n",
    "seqIDs = []\n",
    "for seq_record in SeqIO.parse(\"data/gencode_5utrs_human_cleaned.fa\", \"fasta\"):\n",
    "    seqs.append(str(seq_record.seq))\n",
    "    seqIDs.append(seq_record.id)"
   ],
   "id": "a2a917062050b102",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "seqs = np.vstack((np.array(seqIDs), np.array(seqs))).T",
   "id": "bf735d9ade484e08",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "humangroup = np.array([['Human'] * len(seqs)])",
   "id": "3ee6685882ff3229",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "seqs = np.concatenate((seqs, humangroup.T), axis=1)#!!!!",
   "id": "e3fd3628e273a376",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "for i in range(len(seqs)):\n",
    "    seqs[i, 1] = seqs[i, 1].replace('T', 'U')"
   ],
   "id": "e7a0ba031da1b80d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "human160seqs=humanseqs[humanseqs[\"Sequence Length\"] >= 160]\n",
    "for i in range(len(human160seqs['Sequence'])):\n",
    "    seq = human160seqs.iloc[i, 0]\n",
    "    human160seqs.iloc[i, 0] = seq[len(seq) - 160: len(seq)]\n",
    "human160seqs.loc[:, 'Sequence Length'] = human160seqs['Sequence'].apply(len)"
   ],
   "id": "97c9f0967a97111b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "secstructs=[]\n",
    "mfes=[]\n",
    "for seq in human160seqs[\"Sequence\"]:\n",
    "    (ss,mfe)=vrna.fold(seq)\n",
    "    secstructs.append(ss)\n",
    "    mfes.append(mfe)"
   ],
   "id": "cd1351e945e1ecd5",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "human160seqs.loc[:,'MFE']=mfes\n",
    "human160seqs.loc[:,'ss']=secstructs"
   ],
   "id": "191bf601d5d830ef",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "human160seqs.loc[:,['Sequence','Group','Sequence Length','MFE','ss']]",
   "id": "d9d6cc73f3513083",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "human160seqs.loc[:,['Sequence','Group','Sequence Length','MFE','ss']].to_csv('/fast/AG_Ohler/jdemoli/bachelorgit/data/MFE_human160_stats.csv')",
   "id": "843fe74e72e32cb0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-11-11T15:39:22.743277Z",
     "start_time": "2024-11-11T15:39:18.267642Z"
    }
   },
   "cell_type": "code",
   "source": [
    "'''In previous Iterations of the code we have created a dataset which contains all of our sequences and their relevant data. This dataset is now loaded while we skip the steps before this'''\n",
    "seqs=pd.read_csv('data/new_dataset.csv')\n",
    "seqs.set_index('SeqID',inplace=True)\n",
    "seqs"
   ],
   "id": "1c9b2cea04ca2c25",
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "[Errno 9] Bad file descriptor",
     "output_type": "error",
     "traceback": [
      "\u001B[0;31m---------------------------------------------------------------------------\u001B[0m",
      "\u001B[0;31mOSError\u001B[0m                                   Traceback (most recent call last)",
      "Cell \u001B[0;32mIn[5], line 2\u001B[0m\n\u001B[1;32m      1\u001B[0m \u001B[38;5;124;03m'''In previous Iterations of the code we have created a dataset which contains all of our sequences and their relevant data. This dataset is now loaded while we skip the steps before this'''\u001B[39;00m\n\u001B[0;32m----> 2\u001B[0m seqs\u001B[38;5;241m=\u001B[39m\u001B[43mpd\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mread_csv\u001B[49m\u001B[43m(\u001B[49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[38;5;124;43mdata/new_dataset.csv\u001B[39;49m\u001B[38;5;124;43m'\u001B[39;49m\u001B[43m)\u001B[49m\n\u001B[1;32m      3\u001B[0m seqs\u001B[38;5;241m.\u001B[39mset_index(\u001B[38;5;124m'\u001B[39m\u001B[38;5;124mSeqID\u001B[39m\u001B[38;5;124m'\u001B[39m,inplace\u001B[38;5;241m=\u001B[39m\u001B[38;5;28;01mTrue\u001B[39;00m)\n\u001B[1;32m      4\u001B[0m seqs\n",
      "File \u001B[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1026\u001B[0m, in \u001B[0;36mread_csv\u001B[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001B[0m\n\u001B[1;32m   1013\u001B[0m kwds_defaults \u001B[38;5;241m=\u001B[39m _refine_defaults_read(\n\u001B[1;32m   1014\u001B[0m     dialect,\n\u001B[1;32m   1015\u001B[0m     delimiter,\n\u001B[0;32m   (...)\u001B[0m\n\u001B[1;32m   1022\u001B[0m     dtype_backend\u001B[38;5;241m=\u001B[39mdtype_backend,\n\u001B[1;32m   1023\u001B[0m )\n\u001B[1;32m   1024\u001B[0m kwds\u001B[38;5;241m.\u001B[39mupdate(kwds_defaults)\n\u001B[0;32m-> 1026\u001B[0m \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[43m_read\u001B[49m\u001B[43m(\u001B[49m\u001B[43mfilepath_or_buffer\u001B[49m\u001B[43m,\u001B[49m\u001B[43m \u001B[49m\u001B[43mkwds\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/io/parsers/readers.py:626\u001B[0m, in \u001B[0;36m_read\u001B[0;34m(filepath_or_buffer, kwds)\u001B[0m\n\u001B[1;32m    623\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m parser\n\u001B[1;32m    625\u001B[0m \u001B[38;5;28;01mwith\u001B[39;00m parser:\n\u001B[0;32m--> 626\u001B[0m     \u001B[38;5;28;01mreturn\u001B[39;00m parser\u001B[38;5;241m.\u001B[39mread(nrows)\n",
      "File \u001B[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1996\u001B[0m, in \u001B[0;36mTextFileReader.__exit__\u001B[0;34m(self, exc_type, exc_value, traceback)\u001B[0m\n\u001B[1;32m   1990\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21m__exit__\u001B[39m(\n\u001B[1;32m   1991\u001B[0m     \u001B[38;5;28mself\u001B[39m,\n\u001B[1;32m   1992\u001B[0m     exc_type: \u001B[38;5;28mtype\u001B[39m[\u001B[38;5;167;01mBaseException\u001B[39;00m] \u001B[38;5;241m|\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m   1993\u001B[0m     exc_value: \u001B[38;5;167;01mBaseException\u001B[39;00m \u001B[38;5;241m|\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m   1994\u001B[0m     traceback: TracebackType \u001B[38;5;241m|\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[1;32m   1995\u001B[0m ) \u001B[38;5;241m-\u001B[39m\u001B[38;5;241m>\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[0;32m-> 1996\u001B[0m     \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mclose\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n",
      "File \u001B[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/io/parsers/readers.py:1624\u001B[0m, in \u001B[0;36mTextFileReader.close\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m   1622\u001B[0m \u001B[38;5;28;01mdef\u001B[39;00m \u001B[38;5;21mclose\u001B[39m(\u001B[38;5;28mself\u001B[39m) \u001B[38;5;241m-\u001B[39m\u001B[38;5;241m>\u001B[39m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[1;32m   1623\u001B[0m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mhandles \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[0;32m-> 1624\u001B[0m         \u001B[38;5;28;43mself\u001B[39;49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mhandles\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mclose\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m   1625\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39m_engine\u001B[38;5;241m.\u001B[39mclose()\n",
      "File \u001B[0;32m/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/pandas/io/common.py:144\u001B[0m, in \u001B[0;36mIOHandles.close\u001B[0;34m(self)\u001B[0m\n\u001B[1;32m    142\u001B[0m     \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mcreated_handles\u001B[38;5;241m.\u001B[39mremove(\u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mhandle)\n\u001B[1;32m    143\u001B[0m \u001B[38;5;28;01mfor\u001B[39;00m handle \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mcreated_handles:\n\u001B[0;32m--> 144\u001B[0m     \u001B[43mhandle\u001B[49m\u001B[38;5;241;43m.\u001B[39;49m\u001B[43mclose\u001B[49m\u001B[43m(\u001B[49m\u001B[43m)\u001B[49m\n\u001B[1;32m    145\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mcreated_handles \u001B[38;5;241m=\u001B[39m []\n\u001B[1;32m    146\u001B[0m \u001B[38;5;28mself\u001B[39m\u001B[38;5;241m.\u001B[39mis_wrapped \u001B[38;5;241m=\u001B[39m \u001B[38;5;28;01mFalse\u001B[39;00m\n",
      "\u001B[0;31mOSError\u001B[0m: [Errno 9] Bad file descriptor"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "cell_type": "code",
   "id": "98ca2f2dcb49f573",
   "metadata": {},
   "source": [
    "'''We split the dataset into it's groups for mor flexibility in plotting '''\n",
    "gc50seqs=seqs.loc[seqs['Group']=='Random_gc50']\n",
    "gc70seqs=seqs.loc[seqs['Group']=='Random_gc70']\n",
    "lengthdistributed=pd.concat([humanseqs,gc50dlseqs])"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "humanseqs=seqs.loc[seqs['Group']=='Human']\n",
    "gc50dlseqs=seqs.loc[seqs['Group']=='Random_dl_gc_adjusted']"
   ],
   "id": "b018b731eb76a672",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7034beffe6cc5589",
   "metadata": {},
   "source": [
    "humanseqs160=humanseqs.loc[humanseqs[\"Sequence Length\"] <= 160]\n",
    "humanseqs160=humanseqs160[humanseqs160[\"Sequence Length\"]>=100]\n",
    "humanseqs100=humanseqs.loc[humanseqs[\"Sequence Length\"]<=100]\n",
    "print(len(humanseqs160)/len(humanseqs))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# fish dataset",
   "id": "c2c186a8e23fbb18"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fishseqs=pd.read_csv('data/fish_fputrs_all_iso.csv')\n",
    "\n",
    "fishseqs"
   ],
   "id": "871d64fc06cdff52",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs.drop(['Unnamed: 0','gene_id'],axis=1,inplace=True)\n",
   "id": "55f735fd0f16ce0d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fishseqs.rename(columns={'transcript_id':'SeqID','fputr':'Sequence','fputr_len':'Sequence Length'}, inplace=True)\n",
    "fishseqs.set_index('SeqID',inplace=True)"
   ],
   "id": "9c3cd510a6ab2836",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs['Group']=['Fish']*len(fishseqs)",
   "id": "f7e377ca8ff92e51",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs['GC Content']=polygraph.sequence.gc(fishseqs)",
   "id": "29637e495b564ed8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs",
   "id": "4f67c0137386381d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "for i in fishseqs.index:\n",
    "    fishseqs.loc[i,'Sequence'] = fishseqs.loc[ i,'Sequence'].replace('T', 'U')\n",
    "    "
   ],
   "id": "ec7f26fbd8cdb945",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs",
   "id": "8fb25ed364450b8c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fishAUGcounts=src.rna_analysis.uORFs(fishseqs,'AUG')\n",
    "fishCUGcounts=src.rna_analysis.uORFs(fishseqs,'CUG')\n",
    "fishGUGcounts=src.rna_analysis.uORFs(fishseqs,'GUG')\n",
    "fishUUGcounts=src.rna_analysis.uORFs(fishseqs,'UUG')\n",
    "fishACGcounts=src.rna_analysis.uORFs(fishseqs,'ACG')\n",
    "fishcounts=src.rna_analysis.counts_concat_all(fishseqs,fishAUGcounts,fishCUGcounts,fishGUGcounts,fishUUGcounts,fishACGcounts)\n",
    "fishcounts\n"
   ],
   "id": "31b4f3f0241cc202",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs=pd.merge(fishseqs,fishcounts,on='SeqID',how='left')",
   "id": "24a1d1534c537dc5",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs.to_csv('data/new_fish_uORFstats.csv')",
   "id": "31ad4c86007c1bc4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "old_fishseqs=pd.read_csv('data/fish_uORFstats.csv')\n",
    "old_fishseqs.set_index('SeqID',inplace=True)\n",
    "old_fishseqs"
   ],
   "id": "55eb813cacfcfbda",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "for i in fishseqs.index:\n",
    "    fishseqs.loc[i,'Group'] = fishseqs.loc[ i,'Group'].replace('Fish', 'Old Fish')"
   ],
   "id": "126d314065ce180c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "fishseqs=pd.read_csv('data/new_fish_uORFstats.csv')",
   "id": "e28a9b557b195127",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# yeast dataset",
   "id": "707999b06b7a3e16"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-06T13:37:01.853946Z",
     "start_time": "2025-01-06T13:37:01.782046Z"
    }
   },
   "cell_type": "code",
   "source": "yeastseqs=polygraph.input.read_seqs(\"data/yeastseqs.txt\", incl_ids=True )",
   "id": "5415652f43ff1a96",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-06T13:27:33.207077Z",
     "start_time": "2025-01-06T13:27:32.758494Z"
    }
   },
   "cell_type": "code",
   "source": [
    "gtf= pr.read_gtf(\"data/Saccharomyces_cerevisiae.R64-1-1.113.gtf\")\n",
    "gtf_df= gtf.df\n",
    "fputrs=gtf_df[gtf_df['Feature']=='five_prime_utr']"
   ],
   "id": "e6935b77bf870f9b",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-06T13:50:19.511899Z",
     "start_time": "2025-01-06T13:50:19.449767Z"
    }
   },
   "cell_type": "code",
   "source": [
    "seqIDs=[]\n",
    "for seq_record in SeqIO.parse(\"data/Saccharomyces_cerevisiae.R64-1-1.dna.toplevel.fa\", \"fasta\"):\n",
    "    seqIDs.append(seq_record.id)\n",
    "seqIDs"
   ],
   "id": "55ac0c004c2ad62d",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['I',\n",
       " 'II',\n",
       " 'III',\n",
       " 'IV',\n",
       " 'V',\n",
       " 'VI',\n",
       " 'VII',\n",
       " 'VIII',\n",
       " 'IX',\n",
       " 'X',\n",
       " 'XI',\n",
       " 'XII',\n",
       " 'XIII',\n",
       " 'XIV',\n",
       " 'XV',\n",
       " 'XVI',\n",
       " 'Mito']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-01-06T14:23:04.144480Z",
     "start_time": "2025-01-06T14:23:04.141335Z"
    }
   },
   "cell_type": "code",
   "source": [
    "print(gtf_df.columns)\n",
    "print(gtf_df[\"Feature\"].unique())"
   ],
   "id": "c4e2620f3167916c",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Chromosome', 'Source', 'Feature', 'Start', 'End', 'Score', 'Strand',\n",
      "       'Frame', 'gene_id', 'gene_name', 'gene_source', 'gene_biotype',\n",
      "       'transcript_id', 'transcript_name', 'transcript_source',\n",
      "       'transcript_biotype', 'tag', 'exon_number', 'exon_id', 'protein_id'],\n",
      "      dtype='object')\n",
      "['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon', 'five_prime_utr']\n",
      "Categories (7, object): ['CDS', 'exon', 'five_prime_utr', 'gene', 'start_codon', 'stop_codon', 'transcript']\n"
     ]
    }
   ],
   "execution_count": 16
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Plots",
   "id": "4fa53f3b5b45d972"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "sns.histplot(human160seqs,x='MFE')",
   "id": "b1ae8ac69a1642ea",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "#Sequenzlaenge\n",
    "sns.violinplot(data=seqs,x='Group',y='Sequence Length');"
   ],
   "id": "e3887879a0810911",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "%matplotlib inline\n",
    "sns.violinplot(data=fishseqs,x='Group',y='GC Content')\n",
    "sns.violinplot(data=old_fishseqs,x='Group',y='GC Content')\n",
    "#sns.violinplot(data=humanseqs,x='Group',y='GC Content')\n",
    "#sns.violinplot(data=gc50dlseqs,x='Group',y='GC Content')\n",
    "plt.show()"
   ],
   "id": "cfd9509c75ff3d07",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "old_fishseqs[\"GC Content\"].median()",
   "id": "d5ef80a47ee4d077",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "Í#GC Content \n",
    "%matplotlib \n",
    "sns.violinplot(data=seqs,x='Group',y='GC Content');\n",
    "plt.show()"
   ],
   "id": "4d06b4400af9e9ca",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# uORF Anzahl(Summe aus allen reading frames)\n",
    "sns.violinplot(data=seqs,x='Group',y='all uORFs');\n",
    "plt.ylim(0,100)"
   ],
   "id": "b25c5de704707bee",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "#GC Content vs. uORF anzahl\n",
    "fig,axs = plt.subplots(nrows=1, ncols=2,layout='constrained' )\n",
    "im1= axs[0].hist2d(gc50seqs['GC Content'],gc50seqs['all uORFs'],norm=colors.LogNorm(),cmap='inferno',bins=10)\n",
    "im2= axs[1].hist2d(gc70seqs['GC Content'],gc70seqs['all uORFs'],norm=colors.LogNorm(),cmap='inferno',bins=8)\n",
    "axs[0].set_title(\"50% GC Content\")\n",
    "axs[1].set_title(\"70% GC Content\")\n",
    "#fig.colorbar(im1[3],ax=axs[0])\n",
    "#fig.colorbar(im2[3],ax=axs[1])\n",
    "fig.suptitle('GC Content vs uORF counts',fontsize=14)\n",
    "plt.show()"
   ],
   "id": "4bde7a2f002c2a06",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "    #Menge an overlapping uORFs pro Gruppe\n",
    "im1=sns.histplot(x=seqs[\"AUG_ouORFs\"],hue=seqs[\"Group\"],multiple=\"dodge\",binwidth=0.25)\n",
    "im1.set_yscale('log')\n",
    "im1.set_xticks([0,1,2,3])\n",
    "im1.set_xlabel(\"ouORFs\")\n",
    "plt.title('ouORFs by Group')\n",
    "plt.show()"
   ],
   "id": "8df5fea333c71b7",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "#overlapping uORFs in allen reading Frames\n",
    "fig2, axs2=plt.subplots(1,2,layout='constrained',sharey=True,sharex=True,figsize=(10,5) )\n",
    "sns.histplot(x=humanseqs[\"GUG_uORFs\"],multiple=\"dodge\",ax=axs2[0],bins=30)\n",
    "sns.histplot(x=humanseqs[\"GUG_frame3_uORFs\"],multiple=\"dodge\",ax=axs2[1],legend=False,bins=20,label='Frame 3')\n",
    "sns.histplot(x=humanseqs[\"GUG_frame2_uORFs\"],multiple=\"dodge\",ax=axs2[1],legend=False,bins=20,label='Frame 2')\n",
    "sns.histplot(x=humanseqs[\"GUG_frame1_uORFs\"],multiple=\"dodge\",ax=axs2[1],legend=False,bins=20,label='Frame 1')\n",
    "for ax in axs2:\n",
    "        ax.set_yscale('log')\n",
    "        ax.set_xlabel(\"\")\n",
    "axs2[0].set_title(\"Alle Frames\")\n",
    "axs2[1].set_title(\"Einzelne Frames\")\n",
    "\n",
    "plt.legend(loc='upper right')\n",
    "fig2.suptitle('uORFs in all reading frames',fontsize=14)\n",
    "plt.show()"
   ],
   "id": "4f5e56fc980dbc7a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "- Frame 1 has more uORFs then 2 and 3 -> possible necessity/use of uORFs?\n",
    "- similar phenomenon is not observed with CUG uORFs\n",
    "- slightly more with GUG uORFs"
   ],
   "id": "79b539f8ac34308"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fig3, axs3 =plt.subplots(2,2,layout='tight',sharey=True,figsize=(10,5.5))\n",
    "sns.histplot(x=gc50seqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[0,0])\n",
    "sns.histplot(x=gc70seqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[0,1])\n",
    "sns.histplot(x=humanseqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[1,0])\n",
    "sns.histplot(x=gc50dlseqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[1,1])\n",
    "for arr in axs3:\n",
    "    for ax in arr:\n",
    "        ax.minorticks_on()\n",
    "        ax.set_xlabel(\"\")\n",
    "        #ax.set_xlim(0,100)\n",
    "        #ax.set_yscale('log')\n",
    "axs3[0,0].set_title(\"Random GC=50\")\n",
    "axs3[0,1].set_title(\"Random GC=70\")\n",
    "axs3[1,0].set_title(\"Human\")\n",
    "axs3[1,1].set_title(\"Random GC=50 - sequence length distributed\")\n",
    "fig3.suptitle(\"Length distribution of uORFs\",fontsize=14)\n",
    "plt.show()"
   ],
   "id": "6d9d421fd65d60c1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "a01b69c6974496f9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "- higher GC Content: More ouORFs, less full uORFs",
   "id": "d78c9f7145d06d3d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "hue_order=['Random_dl_gc50','Human']\n",
    "sns.histplot(data=lengthdistributed,stat='probability', x='GUG_uORFs',common_norm=False, hue='Group',hue_order=hue_order)\n",
    "plt.xscale('log')\n",
    "#plt.yscale('log')\n",
    "#plt.xticks(np.arange(1,31));\n",
    "#plt.ylim(0.01,10000)\n",
    "plt.title(\"uORF count distribution\")\n",
    "plt.show()\n",
    "\n"
   ],
   "id": "c4d32a4e6666c0ca",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "|- Human sequences are depleted of AUG uORFs(over 50% have none)\n",
    "- Less depleted in GUG uORFs(close to 40%)\n",
    "- even slgihtly less in CUG uORFs( a bit over 30%)\n",
    "- Random Sequences at 27%"
   ],
   "id": "cd0566d994d9676d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "hue_order=['Random_dl_gc50','Human']\n",
    "sns.kdeplot(data=lengthdistributed, x='Sequence Length', hue='Group',hue_order=hue_order,log_scale=True)\n",
    "plt.title(\"Sequence Length distribution\")\n",
    "plt.show()"
   ],
   "id": "da2693b97f1f16de",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fig4,axs4=plt.subplots(2,2,layout='constrained',sharex=True,sharey=True,figsize=(10,5))\n",
    "\n",
    "sns.histplot(data=lengthdistributed,x=\"all uORFs\",hue=\"Group\",stat='percent',ax=axs4[0,0],bins=200,hue_order=hue_order,common_norm=False)\n",
    "sns.histplot(lengthdistributed,x=\"AUG_uORFs\",hue=\"Group\",stat='percent',ax=axs4[0,1],bins=200,hue_order=hue_order,common_norm=False)\n",
    "sns.histplot(lengthdistributed,x=\"CUG_uORFs\",hue=\"Group\",stat='percent',ax=axs4[1,0],bins=200,hue_order=hue_order,common_norm=False)\n",
    "sns.histplot(lengthdistributed,x=\"GUG_uORFs\",hue=\"Group\",stat='percent',ax=axs4[1,1],bins=200,hue_order=hue_order,common_norm=False)\n",
    "for arr in axs4:\n",
    "    for ax in arr:\n",
    "        ax.set_xlim(0,100)\n",
    "    \n",
    "plt.show()"
   ],
   "id": "fca87535179f0ec3",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "#MFE\n",
    "fig5, axs5 = plt.subplots(1,2,layout='constrained',sharex=True,sharey=True,figsize=(10,5))\n",
    "sns.scatterplot(humanseqs,x=\"Sequence Length\",y=\"MFE\",ax=axs5[0],label='Human')\n",
    "sns.scatterplot(gc50dlseqs,x=\"Sequence Length\",y=\"MFE\",ax=axs5[1],label='Random')\n",
    "plt.savefig('data/MFE.png')\n"
   ],
   "id": "16fd681ab0938bba",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "946b7ef55630ef0e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fig6, axs6 = plt.subplots(1,2,layout='constrained',sharex=True,sharey=True,figsize=(10,5))\n",
    "sns.histplot(x=humanseqs[\"MFE\"]/humanseqs[\"Sequence Length\"],ax=axs6[0])\n",
    "sns.histplot(x=gc50dlseqs[\"MFE\"]/gc50dlseqs[\"Sequence Length\"],ax=axs6[1])\n",
    "for ax in axs6:\n",
    "    ax.set_xlabel('MFE per bp')\n",
    "plt.savefig('data/MFE_per_bp.png')"
   ],
   "id": "f099284b19011893",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# Metadata",
   "id": "409475d2463028a3"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "'''Plot of runtimes of the rna_analysis method over the size of the data (number or length of sequences)'''\n",
    "times=[0.212,1.920,19.402,200.295]\n",
    "sizes=[100,1000,10000,100000]\n",
    "plt.plot(sizes,times)\n",
    "plt.xlabel(\"Size of the data\")\n",
    "plt.ylabel(\"Time [s]\")\n",
    "plt.title(\"runtime of oORFs algorithm vs size of data\")\n",
    "plt.show()"
   ],
   "id": "bddc84b6c672c546",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "times=[0.021,0.256,37.857,3753]\n",
    "sizes=[10,100,1000,10000]\n",
    "plt.plot(sizes,times)\n",
    "plt.xlabel(\"Size of the data\")\n",
    "plt.ylabel(\"Time [s]\")\n",
    "plt.title(\"runtime of outdated randomlengthseqs algorithm vs size of data\")\n",
    "plt.show()"
   ],
   "id": "acab16779abe6c80",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "# testing  ",
   "id": "82162840b11f58b2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def testfunc(a,b):\n",
    "    print(a+b)\n",
    "def testingtestfunc(func,a,b):\n",
    "    func(a,b)\n",
    "    \n",
    "testingtestfunc(testfunc,1,2)"
   ],
   "id": "32ca32eb3abb208a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "print(sum)",
   "id": "7f022bac03543c9d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "cleanedhumanseqs=src.human_sequences.delDupes(cleanedhumanseqs)\n",
    "#np.savetxt(\"data/humanseqsRNA.txt\",humanseqs,delimiter=\"\\t\",newline=\"\\n\",fmt='%s')"
   ],
   "id": "7639479d4bd8ae9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "rng = np.random.default_rng(seed=3)\n",
    "a=rng.choice([0,1,3],3)\n",
    "b=np.array([0,1,2,3])\n",
    "for i in b[1:]:\n",
    "    print(i)"
   ],
   "id": "ccbd1e77a846ab32",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def lognormalfunc(x,myu,sigma,a):\n",
    "    return a*np.exp(-1*(np.log(x)-myu)**2/(2*sigma)**2)/(x*sigma*np.sqrt(2*np.pi))"
   ],
   "id": "c1be1d26771292cb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "sns.scatterplot(x=x_plot,y=lognormalfunc(x_plot,myu,sigma,factor_a),size=1,marker=\".\");",
   "id": "b6371a721e871c65",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "x_plot=np.arange(1,8000)\n",
    "distribution=lognormalfunc(x_plot,myu,sigma,factor_a)\n",
    "distribution=distribution/sum(distribution)\n",
    "#print(sum(distribution))\n",
    "rng=np.random.default_rng()\n",
    "lengths=rng.choice(x_plot, 100000 , p=distribution)"
   ],
   "id": "3622c3b50dd7cf8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "im3=sns.histplot(lengths,kde=True)\n",
    "im3.lines[0].set_color('crimson')\n"
   ],
   "id": "57529f7f0c5b16d4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "im4=sns.scatterplot(x=x_plot,y=distribution,size=1,marker=\".\")\n",
    "#im4.lines[0].set_color('crimson')"
   ],
   "id": "9242eb4f848c4b8d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "lengthdata=list(humanseqs['Sequence Length'])\n",
    "lengthdata=Counter(lengthdata)\n",
    "x,y=[],[]\n",
    "for i,j in lengthdata.items():\n",
    "    x.append(i)\n",
    "    y.append(j)\n",
    "y=np.array(y)\n",
    "sns.scatterplot(x=x,y=y,size=1,marker=\".\");"
   ],
   "id": "26826a8230cd62c2",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "params, cov_matrix= curve_fit(lognormalfunc,x,y)\n",
    "x_plot=np.arange(1,8000)\n",
    "plt.plot(x_plot,lognormalfunc(x_plot,params[0],params[1],params[2]))\n",
    "plt.plot(x,y,linestyle='None',marker=\".\")\n",
    "\n",
    "myu=params[0]\n",
    "sigma=params[1]\n",
    "factor_a=params[2]\n",
    "print(myu,sigma,factor_a)\n"
   ],
   "id": "45b5c03cb6609d49",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def poissonfunc(x,lamb,a):\n",
    "    return a*poisson.pmf(x,lamb)\n",
    "params1, cov_matrix1 = curve_fit(poissonfunc,x,y)\n",
    "x_plot1=(np.arange(0,8000))\n",
    "plt.plot(x_plot1,poissonfunc(x_plot1,params1[0],params1[1]))\n",
    "plt.plot(x,y,linestyle='None',marker=\".\")"
   ],
   "id": "46f47d2d9f3cd293",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "x_plot=np.arange(1,8000)\n",
    "def softmax(x):\n",
    "    return np.exp(x)/sum(np.exp(x))\n",
    "\n",
    "    \n",
    "#plt.plot(x_plot,lognormalfunc(x_plot,myu,sigma))\n",
    "#plt.plot(x_plot,softmax(lognormalfunc(x_plot,myu,sigma)))\n",
    "#plt.plot(x_plot,scipy.special.softmax(lognormalfunc(x_plot,myu,sigma)))\n",
    "print([sum(lognormalfunc(x_plot,myu,sigma)),sum(softmax(lognormalfunc(x_plot,myu,sigma)))])"
   ],
   "id": "3370ab8c3217201c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "#New approach for uORFs function\n",
    "def uORFs(seqs):\n",
    "    frames=src.rna_analysis.codons(seqs)\n",
    "    template={\n",
    "        \"uORFs\":list(np.zeros(len(frames))),\n",
    "        \"ouORFs\":list(np.zeros(len(frames))),\n",
    "        \"mean_uORF_length\":list(np.zeros(len(frames))),\n",
    "        \n",
    "        \"CUG_uORFs\":list(np.zeros(len(frames))),\n",
    "        \"CUG_ouORFs\":list(np.zeros(len(frames))),\n",
    "        \"mean_CUG_uORF_length\":list(np.zeros(len(frames))),\n",
    "        \n",
    "        \n",
    "        \"GUG_uORFs\":list(np.zeros(len(frames))),\n",
    "        \"GUG_ouORFs\":list(np.zeros(len(frames))),\n",
    "        \"mean_GUG_uORF_length\":list(np.zeros(len(frames))),\n",
    "    }\n",
    "    framedict={}\n",
    "    for frame in range(3):\n",
    "        framedict[f\"frame{frame+1}\"]=pd.DataFrame(template)\n",
    "        for index in range(len(frames[f\"frame{frame+1}\"])):\n",
    "            seq=frames[f\"frame{frame+1}\"][index]\n",
    "            lengths=[]\n",
    "            CUG_lengths=[]\n",
    "            GUG_lengths=[]\n",
    "            for codon in range(len(seq)):\n",
    "                if seq[codon]=='AUG':\n",
    "                    uORF_length=0\n",
    "                    ouORF_bool=True\n",
    "                    for nextcodon in range(len(seq)-codon-1):\n",
    "                        uORF_length+=3\n",
    "                        if seq[nextcodon+codon+1] in src.constants.stopcodons:\n",
    "                            ouORF_bool=False\n",
    "                            framedict[f\"frame{frame+1}\"].loc[index,\"uORFs\"]+=1\n",
    "                            lengths.append(uORF_length)\n",
    "                            break\n",
    "                    if ouORF_bool:\n",
    "                        framedict[f\"frame{frame+1}\"].loc[index,\"ouORFs\"]+=1\n",
    "                        break\n",
    "                    \n",
    "                if seq[codon]=='CUG':\n",
    "                    uORF_length=0\n",
    "                    ouORF_bool=True\n",
    "                    for nextcodon in range(len(seq)-codon-1):\n",
    "                        uORF_length+=3\n",
    "                        if seq[nextcodon+codon+1] in src.constants.stopcodons:\n",
    "                            ouORF_bool=False\n",
    "                            framedict[f\"frame{frame+1}\"].loc[index,\"CUG_uORFs\"]+=1\n",
    "                            CUG_lengths.append(uORF_length)\n",
    "                            break\n",
    "                    if ouORF_bool:\n",
    "                        framedict[f\"frame{frame+1}\"].loc[index,\"CUG_ouORFs\"]+=1\n",
    "                        break\n",
    "                        \n",
    "                if seq[codon]=='GUG':\n",
    "                    uORF_length=0\n",
    "                    ouORF_bool=True\n",
    "                    for nextcodon in range(len(seq)-codon-1):\n",
    "                        uORF_length+=3\n",
    "                        if seq[nextcodon+codon+1] in src.constants.stopcodons:\n",
    "                            ouORF_bool=False\n",
    "                            framedict[f\"frame{frame+1}\"].loc[index,\"GUG_uORFs\"]+=1\n",
    "                            GUG_lengths.append(uORF_length)\n",
    "                            break\n",
    "                    if ouORF_bool:\n",
    "                        framedict[f\"frame{frame+1}\"].loc[index,\"GUG_ouORFs\"]+=1\n",
    "                        break\n",
    "             \n",
    "            if len(lengths)>0:\n",
    "                framedict[f\"frame{frame+1}\"].loc[index,\"mean_uORF_length\"] = sum(lengths)/len(lengths)   \n",
    "            \n",
    "            if len(CUG_lengths)>0:\n",
    "                framedict[f\"frame{frame+1}\"].loc[index,\"mean_CUG_uORF_length\"]=sum(CUG_lengths)/len(CUG_lengths)\n",
    "            if len(GUG_lengths)>0:\n",
    "                framedict[f\"frame{frame+1}\"].loc[index,\"mean_GUG_uORF_length\"]=sum(GUG_lengths)/len(GUG_lengths)\n",
    "             \n",
    "    counts=pd.DataFrame()\n",
    "    \n",
    "    for key in framedict:\n",
    "        for column in framedict[key].columns:\n",
    "            counts[f\"{key}_{column}\"] = framedict[key].loc[:,column]\n",
    "            \n",
    "    counts[\"uORFs\"]=counts[\"frame1_uORFs\"]+counts[\"frame2_uORFs\"]+counts[\"frame3_uORFs\"]\n",
    "    counts[\"CUG_uORFs\"]=counts[\"frame1_CUG_uORFs\"]+counts[\"frame2_CUG_uORFs\"]+counts[\"frame3_CUG_uORFs\"]\n",
    "    counts[\"GUG_uORFs\"]=counts[\"frame1_GUG_uORFs\"]+counts[\"frame2_GUG_uORFs\"]+counts[\"frame3_GUG_uORFs\"]\n",
    "    counts[\"ouORFs\"]=counts[\"frame1_ouORFs\"]+counts[\"frame2_ouORFs\"]+counts[\"frame3_ouORFs\"]\n",
    "    counts[\"CUG_ouORFs\"]=counts[\"frame1_CUG_ouORFs\"]+counts[\"frame2_CUG_ouORFs\"]+counts[\"frame3_CUG_ouORFs\"]\n",
    "    counts[\"GUG_ouORFs\"]=counts[\"frame1_GUG_ouORFs\"]+counts[\"frame2_GUG_ouORFs\"]+counts[\"frame3_CUG_ouORFs\"]\n",
    "    counts[\"all uORFs\"]=counts[\"uORFs\"]+counts[\"CUG_uORFs\"]+counts[\"GUG_uORFs\"]\n",
    "    counts[\"all ouORFs\"]=counts[\"ouORFs\"]+counts[\"CUG_ouORFs\"]+counts[\"GUG_ouORFs\"]\n",
    "    \n",
    "    counts = counts.replace(0, np.nan)\n",
    "    counts[\"mean_uORF_length\"] = counts.loc[:,[\"frame1_mean_uORF_length\",\"frame2_mean_uORF_length\",\"frame3_mean_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"mean_CUG_uORF_length\"] = counts.loc[:,[\"frame1_mean_CUG_uORF_length\",\"frame2_mean_CUG_uORF_length\",\"frame3_mean_CUG_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"mean_GUG_uORF_length\"] = counts.loc[:,[\"frame1_mean_GUG_uORF_length\",\"frame2_mean_GUG_uORF_length\",\"frame3_mean_GUG_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"all mean lengths\"] = counts.loc[:,[\"mean_uORF_length\",\"mean_CUG_uORF_length\",\"mean_GUG_uORF_length\"]].mean(axis=1)\n",
    "    counts = counts.replace(np.nan, 0)\n",
    "\n",
    "    cols_in_front = [\"all uORFs\",\"all ouORFs\",\"all mean lengths\",\"uORFs\",\"CUG_uORFs\",\"GUG_uORFs\",\"ouORFs\",\"CUG_ouORFs\",\"GUG_ouORFs\",\"mean_uORF_length\",\"mean_CUG_uORF_length\",\"mean_GUG_uORF_length\"]\n",
    "    counts = counts.loc[:,[c for c in cols_in_front if c in counts.columns] + [c for c in counts if c not in cols_in_front]]\n",
    "    \n",
    "    if type(seqs) == pd.core.frame.DataFrame:\n",
    "        counts.insert(loc=0,column='SeqID',value=seqs.index)\n",
    "        counts.set_index('SeqID', inplace=True)\n",
    "    \n",
    "    return counts"
   ],
   "id": "261e4e29a9dc30f0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "singleseq=\"CACCCCGCUCGCGCCCAGCAGCGGACAUGCUCCACGAGCGCUGACAGCGUCGCCGUGCAACAAUCCUGGCGCCUGCCGCCGGCGAGCGGUCGACCGACCCGCACCGGUUCUCUAUGCCCCGCUACGGGCGGCCCGAGGCUUUACUACCCGGGCGCUUACC\"\n",
    "singleseq1=\"CCCGGCCGCGUCUGACUGCGCACCCCGCAAGUGUGUUGUCCCGCUCCCCAGCAGCCCCCCCCCCCCCCCGCUCGGCGCAGGGCCGAUCGCCACUUCCCUACCCGCGCGCCCACUUAUAGUCCCCGACUACCGUGCGCGGUUACUGCAGAAUGUCCGCCCCU\"\n",
    "#singleseqcodons=src.rna_analysis.codons(singleseq1)\n",
    "counts=src.rna_analysis.uORFs(singleseq1)\n",
    "counts.T\n",
    "\n"
   ],
   "id": "31dc884642c12fff",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "singleseqcodons.T.to_csv(\"data/testcodons.csv\",sep=\"\\t\",index=False)\n",
   "id": "e0dede270c603401",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "504acc582dcf792a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "mydict={\"frame1\":[10,11,12],\"frame2\":[20,21,22],\"frame3\":[30,31,32]}\n",
    "num=69\n",
    "df=pd.DataFrame(mydict)\n",
    "df2=pd.DataFrame()\n",
    "df[\"mean\"]=df.loc[:,[\"frame2\",\"frame3\"]].mean(axis=1)\n",
    "df"
   ],
   "id": "af199c5a23750d7b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "mydict1={\"feature1\":[0],\"feature2\":[0],\"feature3\":[0]}\n",
    "template=pd.DataFrame(mydict1)\n",
    "l=[]\n",
    "for frame in range(3):\n",
    "    r1=np.random.randint(1,10)\n",
    "    r2=np.random.randint(1,10)\n",
    "    r3=np.random.randint(1,10)\n",
    "    df=pd.DataFrame(mydict1)\n",
    "    df.loc[0,\"feature1\"]=r1\n",
    "    df.loc[0,\"feature2\"]=r2\n",
    "    df.loc[0,\"feature3\"]=r3\n",
    "    l.append(df)\n",
    "print(l[0])\n",
    "print(l[1])\n",
    "print(l[2])"
   ],
   "id": "d69d02cebafda0c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "var1=10\n",
    "var2=var1\n",
    "var1=11\n",
    "var2"
   ],
   "id": "40c07afb654c4dce",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "for i in range(10):\n",
    "    print(i)\n",
    "    if i ==5 :\n",
    "        print(\"test\")\n",
    "        for j in range(3):\n",
    "            print(\"test\", 2)\n",
    "            if j==1:\n",
    "                print(\"break\")\n",
    "                break"
   ],
   "id": "38000d91f2241bc4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "singleseq2=\"AGCGCAUGUAUGUAUGCUGCGCGC\"\n",
    "count=0\n",
    "streaks=[]\n",
    "streak=0\n",
    "streakstart=False\n",
    "for i in range(0,len(singleseq2),2):\n",
    "    print(singleseq2[i:i+2])\n",
    "    if singleseq2[i:i+2]==\"GC\" or singleseq2[i:i+2]==\"CG\":\n",
    "        count+=1\n",
    "        if i>=2 and singleseq2[i-2:i]==singleseq2[i:i+2]:\n",
    "            streak+=2\n",
    "    else:\n",
    "        if streak > 0:\n",
    "            streaks.append(streak)\n",
    "            streak=0\n",
    "for i in range(1,len(singleseq2),2):\n",
    "    print(singleseq2[i:i+2])\n",
    "    if singleseq2[i:i+2]==\"GC\" or singleseq2[i:i+2]==\"CG\":\n",
    "        count+=1\n",
    "        if i>=2 and singleseq2[i-2:i]==singleseq2[i:i+2]:\n",
    "            streak+=2\n",
    "    else:\n",
    "        if streak > 0:\n",
    "            streaks.append(streak)\n",
    "            streak=0\n",
    "\n",
    "print(count,streaks)\n",
    "    \n",
    "        \n",
    "        "
   ],
   "id": "8781eebfed8cc188",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def fast_uORfs(seqs,startcodon):\n",
    "    \n",
    "    seq_pos={\"frame1\":[],\"frame2\":[],\"frame3\":[]}\n",
    "    if type(seqs)== pd.DataFrame:\n",
    "        seqstrings=seqs[\"Sequence\"].values\n",
    "    if type(seqs)== str:\n",
    "        seqstrings=[seqs]\n",
    "    len_loop=len(seqstrings)\n",
    "    template = {\n",
    "        \"uORFs\": list(np.zeros(len_loop)),\n",
    "        \"ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"max_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \n",
    "        \"frame1_uORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame1_ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame1_mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"frame1_max_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \n",
    "        \"frame2_uORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame2_ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame2_mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"frame2_max_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \n",
    "        \"frame3_uORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame3_ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame3_mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"frame3_max_uORF_length\": list(np.zeros(len_loop)),\n",
    "    }\n",
    "    counts=pd.DataFrame(template)\n",
    "    \n",
    "    posdict={\"frame1\":[[] for _ in range(len(seqstrings))],\"frame2\":[[] for _ in range(len(seqstrings))],\"frame3\":[[] for _ in range(len(seqstrings))]}\n",
    "    for index in range(len(seqstrings)):\n",
    "        seq=seqstrings[index]\n",
    "        for i in range(len(seq)):\n",
    "            rev_i=len(seq)-i\n",
    "            frame=(rev_i%3)\n",
    "            if seq[i:i+3]==startcodon:\n",
    "                posdict[f'frame{frame+1}'][index].append(i+1)\n",
    "            if seq[i:i+3]in src.constants.stopcodons:\n",
    "                posdict[f'frame{frame+1}'][index].append((i+1)*-1) \n",
    "\n",
    "    \n",
    "    for frame in range(3):\n",
    "        poslist=posdict[f'frame{frame+1}']\n",
    "        for index in range(len_loop):\n",
    "            pos=poslist[index]\n",
    "            if len(pos)>1:\n",
    "                lengths=[]\n",
    "                if pos[-1]>0:\n",
    "                    counts.loc[index,\"ouORFs\"]+=1\n",
    "                    counts.loc[index,f\"frame{frame+1}_ouORFs\"]+=1\n",
    "                    \n",
    "                for i in range(len(pos)):\n",
    "                    if pos[i]>0:\n",
    "                    \n",
    "                        for j in range(len(pos)-i-1):        \n",
    "                            if pos[i+j+1]<0:\n",
    "                                \n",
    "                                counts.loc[index,\"uORFs\"]+=1\n",
    "                                counts.loc[index,f\"frame{frame+1}_uORFs\"]+=1\n",
    "                                lengths.append((pos[i]+pos[i+j+1])*-1+3)\n",
    "                                break\n",
    "                if len(lengths)>0:\n",
    "                    counts.loc[index,f\"frame{frame+1}_mean_uORF_length\"]=sum(lengths)/len(lengths)\n",
    "                    counts.loc[index,f\"frame{frame+1}_max_uORF_length\"]=max(lengths)\n",
    "                if len(pos)==1 and pos[0]>0:\n",
    "                    counts.loc[index,\"ouORFs\"]+=1\n",
    "                    counts.loc[index,f\"frame{frame+1}_ouORFs\"]+=1\n",
    "                    \n",
    "                    \n",
    "                    \n",
    "    \n",
    "    counts[\"mean_uORF_length\"]=counts.loc[:,[\"frame1_mean_uORF_length\",\"frame2_mean_uORF_length\",\"frame3_mean_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"max_uORF_length\"]=counts.loc[:,[\"frame1_max_uORF_length\",\"frame2_max_uORF_length\",\"frame3_max_uORF_length\"]].max(axis=1)\n",
    "    \n",
    "    return counts\n",
    "                 "
   ],
   "id": "4a06972378fa0ebb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "singleseq=\"AUGGCCGCAUAAAGAUGGAAUAGC\"",
   "id": "49d120eb10d6e9e6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "AUG_counts2=fast_uORfs(seqs,'AUG')",
   "id": "511771a86a5c848b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "CUG_counts2=fast_uORfs(seqs,'CUG')\n",
    "GUG_counts2=fast_uORfs(seqs,'GUG')\n",
    "counts2=src.rna_analysis.counts_concat_all(seqs,AUG_counts2,CUG_counts2,GUG_counts2)\n",
    "counts2"
   ],
   "id": "18018709ad1544f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "counts1",
   "id": "84878300d2bc1816",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "seqstrings[-1]",
   "id": "d1afc5b5002239f8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "l=\"AUGUGAUAAAUGCAGAU\"\n",
    "frame1=[]\n",
    "frame2=[]\n",
    "frame3=[]\n",
    "for i in range(len(l)-2):\n",
    "    rev_i=len(l)-i\n",
    "    if  rev_i%3==0:\n",
    "        frame1.append(l[i]+l[i+1]+l[i+2])\n",
    "    if rev_i %3==1:\n",
    "        frame2.append(l[i]+l[i+1]+l[i+2])\n",
    "    if rev_i %3==2:\n",
    "        frame3.append(l[i]+l[i+1]+l[i+2])\n",
    "print(frame1,'\\n',frame2,'\\n',frame3)        "
   ],
   "id": "c82a75caa76612eb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "l=[3,21,24,2,1,2,0,-1,4]\n",
    "print(max(l))"
   ],
   "id": "cf2935a60f014141",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "l300=np.arange(400)\n",
    "for i in range(4000):\n",
    "    for i in range(len(l300)):\n",
    "        for j in range(len(l300)-1):\n",
    "            a=i+j\n"
   ],
   "id": "5590152ebd727324",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "human160seqs = humanseqs[humanseqs[\"Sequence Length\"] >= 160]\n",
    "human160seqs"
   ],
   "id": "6b7ab7057b6e2f45",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "human160seqs['Sequence_adjusted']=humanseqs['Sequence']",
   "id": "a6d75e82233967fa",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "print(src.constants.DNABASES)",
   "id": "af326f9a44650d10",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df=pd.DataFrame({'a':[23,100,212],\n",
    "                 'b':[3,4,5],\n",
    "                 'c':[6,7,8],})\n",
    "for i in range(len(df)):\n",
    "    print(df.loc[i,'a'])"
   ],
   "id": "fa08b9423be686f1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "f26dfbbb2ded4470",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
