{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T08:59:44.746980Z",
     "start_time": "2024-09-16T08:59:42.721787Z"
    }
   },
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "from Bio import SeqIO\n",
    "import ViennaRNA as vrna\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import colors\n",
    "\n",
    "from collections import Counter\n",
    "from scipy.optimize import curve_fit\n",
    "import scipy.special\n",
    "from scipy.stats import poisson\n",
    "\n",
    "import polygraph.input, polygraph.sequence, polygraph.visualize"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T08:59:45.112158Z",
     "start_time": "2024-09-16T08:59:45.093229Z"
    }
   },
   "cell_type": "code",
   "source": "import src.rng_sequences, src.human_sequences,src.constants,src.uORF_Statistics",
   "id": "f50959a437a2866a",
   "outputs": [],
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "id": "d6eae7238000e4b2",
   "metadata": {},
   "source": [
    " # Generating data files"
   ]
  },
  {
   "cell_type": "code",
   "id": "8b92620741b250c5",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T08:10:10.123264Z",
     "start_time": "2024-09-16T08:10:10.118749Z"
    }
   },
   "source": [
    "'''Defining params'''\n",
    "seqnum=100000\n",
    "seqlen=160\n",
    "myu=5.345801185151601\n",
    "sigma=0.8416777786431029\n",
    "factor_a=66009.38523606668\n",
    "params=[myu,sigma,factor_a]\n",
    "np.random.seed(12345)"
   ],
   "outputs": [],
   "execution_count": 2
  },
  {
   "cell_type": "code",
   "id": "fa84b2d355cd02b8",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T08:10:30.484439Z",
     "start_time": "2024-09-16T08:10:24.366809Z"
    }
   },
   "source": [
    "seqs1 = src.rng_sequences.randomseqs(seqnum,seqlen,group_name=\"Random_gc_adjusted\",seed=12345,gc_content=0.6081)\n",
    "seqs2 = src.rng_sequences.randomseqs(seqnum,seqlen,group_name=\"Random_gc50\",seed=12345)"
   ],
   "outputs": [],
   "execution_count": 3
  },
  {
   "cell_type": "code",
   "id": "789af7770c6de70a",
   "metadata": {},
   "source": [
    "np.savetxt('data/small_dataset.txt' , seqs1, delimiter='\\t',newline='\\n',fmt='%s')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4cbd8b5e0035e7b3",
   "metadata": {},
   "source": [
    "testseqs=polygraph.seqs=polygraph.input.read_seqs('data/small_dataset.txt')\n",
    "testseqs"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "48b1df639ad5802c",
   "metadata": {},
   "source": [
    "seqs3=src.rng_sequences.randomlengthseqs(params,8000,seqnum,\"Random_dl_gc50\",seed=12345)\n",
    "seqs4=src.rng_sequences.randomlengthseqs(params,8000,seqnum,\"Random_dl_gc_adjusted\",seed=12345,gc_content=0.6081)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8683b3308497543c",
   "metadata": {},
   "source": [
    "'''This code is used to generate a dataset with continous IDs and save it in a txt file'''\n",
    "randomseqs=np.concatenate([seqs1,seqs2,seqs3,seqs4],axis=0)\n",
    "randomseqsindex=np.array([f\"seq_{i+1}\" for i in range(len(randomseqs))])\n",
    "randomseqs=np.vstack((randomseqsindex,randomseqs.T)).T\n",
    "np.savetxt(\"data/random_dataset.txt\",randomseqs,delimiter=\"\\t\",newline=\"\\n\",fmt='%s')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "c81cd8f6511e1740",
   "metadata": {},
   "source": [
    "'''Here we read out the human genome sequences and save it in a txt file'''\n",
    "humanseqs=src.human_sequences.readFASTA(\"data/gencode_5utrs_human.fa\")\n",
    "humanseqs=src.human_sequences.delDupes(humanseqs)\n",
    "np.savetxt(\"data/humanseqsRNA.txt\",humanseqs,delimiter=\"\\t\",newline=\"\\n\",fmt='%s')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "756f216526ee31f0",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e06a688d3ebe8993",
   "metadata": {},
   "source": [
    "''' Here we write every Sequence into one txt file'''\n",
    "seqfiles=['data/humanseqsRNA.txt','data/random_dataset.txt']\n",
    "with open('data/full_dataset.txt','w') as outfile:\n",
    "    for fname in seqfiles:\n",
    "        with open(fname) as infile:\n",
    "            for line in infile:\n",
    "                outfile.write(line)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5c0239d53acf1e36",
   "metadata": {},
   "source": [
    "seqs=polygraph.input.read_seqs('data/full_dataset.txt',incl_ids=True)\n",
    "seqs"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "mfe_seqs=pd.read_csv('data/MFE_dl_stats.csv')\n",
    "mfe_seqs.set_index('SeqID',inplace=True)\n",
    "mfe_seqs['Sequence Length']=mfe_seqs.Sequence.apply(len)\n",
    "mfe_seqs"
   ],
   "id": "845b3919b86a16b9",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "63e1273ddfa9072a",
   "metadata": {},
   "source": [
    "# Generating statistics for the Sequences "
   ]
  },
  {
   "cell_type": "code",
   "id": "ed57301afa94683d",
   "metadata": {},
   "source": [
    "'''Sequence lengths'''\n",
    "seqs['Sequence Length']=seqs.Sequence.apply(len)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "495ffbd1ee65595a",
   "metadata": {},
   "source": [
    "'''GC Content'''\n",
    "seqs['GC Content']=polygraph.sequence.gc(seqs)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "''' MFE '''\n",
    "secstructs=[]\n",
    "mfes=[]\n",
    "for seq in seqs[\"Sequence\"].iloc[0:1000]:\n",
    "    (ss,mfe)=vrna.fold(seq)\n",
    "    secstructs.append(ss)\n",
    "    mfes.append(mfe)\n",
    "#seqs['MFE']=mfes\n",
    "#seqs['ss']=secstructs\n"
   ],
   "id": "ba6b69bb818e133d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8162b12eee8b0c8a",
   "metadata": {},
   "source": [
    "'''uORF counts'''\n",
    "AUG_counts1=src.uORF_Statistics.uORFs(seqs,'AUG')\n",
    "CUG_counts1=src.uORF_Statistics.uORFs(seqs,'CUG')\n",
    "GUG_counts1=src.uORF_Statistics.uORFs(seqs,'GUG')\n",
    "UUG_counts1=src.uORF_Statistics.uORFs(seqs,'UUG')\n",
    "ACG_counts1=src.uORF_Statistics.uORFs(seqs,'ACG')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "counts1=src.uORF_Statistics.counts_concat_all(seqs,AUG_counts1,CUG_counts1,GUG_counts1,UUG_counts1,ACG_counts1)\n",
    "counts1.to_csv('data/uORF_stats.csv')"
   ],
   "id": "7690df01df4a1c7d",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "771b1edc4bc44108",
   "metadata": {},
   "source": "seqs=pd.concat([seqs,counts],axis=1)\n",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-05T09:58:28.620106Z",
     "start_time": "2024-09-05T09:58:23.795239Z"
    }
   },
   "cell_type": "code",
   "source": [
    "seqs=pd.read_csv('data/uORF_stats.csv')\n",
    "seqs.set_index('SeqID',inplace=True)\n",
    "seqs"
   ],
   "id": "1921362bbfe892fe",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                                                            Sequence  \\\n",
       "SeqID                                                                  \n",
       "ENST00000641515.2  CCCAGAUCUCUUCAGUUUUUAUGCCUCAUUCUGUGAAAAUUGCUGU...   \n",
       "ENST00000616016.5  GGCGGCGGAGUCUCCCAAGUCCCCGCCGGGCGGGCGCGCGCCAGUG...   \n",
       "ENST00000437963.5  CAGCGCUUGGGGCUCGCGGGCCGCUCCCUCCGCUCGGAAGGGAAAA...   \n",
       "ENST00000342066.8  GCAGAGCCCAGCAGAUCCCUGCGGCGUUCGCGAGGGUGGGACGGGA...   \n",
       "ENST00000338591.8  GGGAGUGAGCGACACAGAGCGGGCCGCCACCGCCGAGCAGCCCUCC...   \n",
       "...                                                              ...   \n",
       "seq_399996         UAGAAACGGCCGCAUCGACUCCGAGCCGACACGCCGUAGAGAGCGG...   \n",
       "seq_399997         CCAGUCCUGGAGGUGCUCAAGCGGCGGUGGAGCACAGCUCGUUCGG...   \n",
       "seq_399998         CUGAGUGGUGUCAGGGCCGUACCCAAUCUACACCUAGUGACGGGGA...   \n",
       "seq_399999         UCGGAAGAUACAAAUUGCCCUGCGGCCUAUAUGGAAAGACAGCCUU...   \n",
       "seq_400000         UCGCGCUACGUCGCGCCUACCGUCCGUCGGCGUGGGCUCUUCGGGG...   \n",
       "\n",
       "                                   Group  Sequence Length  GC Content  \\\n",
       "SeqID                                                                   \n",
       "ENST00000641515.2                  Human               60    0.400000   \n",
       "ENST00000616016.5                  Human              509    0.809430   \n",
       "ENST00000437963.5                  Human               60    0.666667   \n",
       "ENST00000342066.8                  Human               90    0.666667   \n",
       "ENST00000338591.8                  Human              110    0.745455   \n",
       "...                                  ...              ...         ...   \n",
       "seq_399996         Random_dl_gc_adjusted              236    0.546610   \n",
       "seq_399997         Random_dl_gc_adjusted              161    0.627329   \n",
       "seq_399998         Random_dl_gc_adjusted              113    0.592920   \n",
       "seq_399999         Random_dl_gc_adjusted              524    0.559160   \n",
       "seq_400000         Random_dl_gc_adjusted               87    0.701149   \n",
       "\n",
       "                   all uORFs  all ouORFs  all mean uORF lengths  \\\n",
       "SeqID                                                             \n",
       "ENST00000641515.2        3.0         2.0               7.500000   \n",
       "ENST00000616016.5       18.0         6.0             158.306667   \n",
       "ENST00000437963.5        0.0         3.0               0.000000   \n",
       "ENST00000342066.8        2.0         3.0              36.000000   \n",
       "ENST00000338591.8        1.0         1.0              63.000000   \n",
       "...                      ...         ...                    ...   \n",
       "seq_399996               7.0         1.0              45.000000   \n",
       "seq_399997               4.0         4.0              33.500000   \n",
       "seq_399998               6.0         5.0              29.833333   \n",
       "seq_399999              48.0         6.0              51.641905   \n",
       "seq_400000               1.0         2.0              12.000000   \n",
       "\n",
       "                   all max uORF lengths  AUG_uORFs  CUG_uORFs  ...  \\\n",
       "SeqID                                                          ...   \n",
       "ENST00000641515.2                   9.0        0.0        2.0  ...   \n",
       "ENST00000616016.5                 318.0        1.0        8.0  ...   \n",
       "ENST00000437963.5                   0.0        0.0        0.0  ...   \n",
       "ENST00000342066.8                  42.0        0.0        1.0  ...   \n",
       "ENST00000338591.8                  63.0        0.0        0.0  ...   \n",
       "...                                 ...        ...        ...  ...   \n",
       "seq_399996                        108.0        1.0        1.0  ...   \n",
       "seq_399997                         69.0        0.0        1.0  ...   \n",
       "seq_399998                         51.0        0.0        1.0  ...   \n",
       "seq_399999                        216.0        6.0       15.0  ...   \n",
       "seq_400000                         12.0        0.0        0.0  ...   \n",
       "\n",
       "                   ACG_frame1_mean_uORF_length  ACG_frame1_max_uORF_length  \\\n",
       "SeqID                                                                        \n",
       "ENST00000641515.2                          0.0                         0.0   \n",
       "ENST00000616016.5                         18.0                        18.0   \n",
       "ENST00000437963.5                          0.0                         0.0   \n",
       "ENST00000342066.8                          0.0                         0.0   \n",
       "ENST00000338591.8                          0.0                         0.0   \n",
       "...                                        ...                         ...   \n",
       "seq_399996                               108.0                       108.0   \n",
       "seq_399997                                 0.0                         0.0   \n",
       "seq_399998                                 0.0                         0.0   \n",
       "seq_399999                                28.0                        45.0   \n",
       "seq_400000                                 0.0                         0.0   \n",
       "\n",
       "                   ACG_frame2_uORFs  ACG_frame2_ouORFs  \\\n",
       "SeqID                                                    \n",
       "ENST00000641515.2               0.0                0.0   \n",
       "ENST00000616016.5               1.0                1.0   \n",
       "ENST00000437963.5               0.0                0.0   \n",
       "ENST00000342066.8               0.0                0.0   \n",
       "ENST00000338591.8               0.0                0.0   \n",
       "...                             ...                ...   \n",
       "seq_399996                      0.0                1.0   \n",
       "seq_399997                      0.0                1.0   \n",
       "seq_399998                      0.0                0.0   \n",
       "seq_399999                      1.0                1.0   \n",
       "seq_400000                      0.0                0.0   \n",
       "\n",
       "                   ACG_frame2_mean_uORF_length  ACG_frame2_max_uORF_length  \\\n",
       "SeqID                                                                        \n",
       "ENST00000641515.2                          0.0                         0.0   \n",
       "ENST00000616016.5                        297.0                       297.0   \n",
       "ENST00000437963.5                          0.0                         0.0   \n",
       "ENST00000342066.8                          0.0                         0.0   \n",
       "ENST00000338591.8                          0.0                         0.0   \n",
       "...                                        ...                         ...   \n",
       "seq_399996                                 0.0                         0.0   \n",
       "seq_399997                                 0.0                         0.0   \n",
       "seq_399998                                 0.0                         0.0   \n",
       "seq_399999                                60.0                        60.0   \n",
       "seq_400000                                 0.0                         0.0   \n",
       "\n",
       "                   ACG_frame3_uORFs  ACG_frame3_ouORFs  \\\n",
       "SeqID                                                    \n",
       "ENST00000641515.2               0.0                0.0   \n",
       "ENST00000616016.5               1.0                0.0   \n",
       "ENST00000437963.5               0.0                0.0   \n",
       "ENST00000342066.8               1.0                0.0   \n",
       "ENST00000338591.8               0.0                0.0   \n",
       "...                             ...                ...   \n",
       "seq_399996                      1.0                0.0   \n",
       "seq_399997                      0.0                0.0   \n",
       "seq_399998                      1.0                0.0   \n",
       "seq_399999                      9.0                1.0   \n",
       "seq_400000                      0.0                1.0   \n",
       "\n",
       "                   ACG_frame3_mean_uORF_length  ACG_frame3_max_uORF_length  \n",
       "SeqID                                                                       \n",
       "ENST00000641515.2                          0.0                         0.0  \n",
       "ENST00000616016.5                        228.0                       228.0  \n",
       "ENST00000437963.5                          0.0                         0.0  \n",
       "ENST00000342066.8                         42.0                        42.0  \n",
       "ENST00000338591.8                          0.0                         0.0  \n",
       "...                                        ...                         ...  \n",
       "seq_399996                                 9.0                         9.0  \n",
       "seq_399997                                 0.0                         0.0  \n",
       "seq_399998                                12.0                        12.0  \n",
       "seq_399999                                52.0                       102.0  \n",
       "seq_400000                                 0.0                         0.0  \n",
       "\n",
       "[482954 rows x 88 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Group</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>GC Content</th>\n",
       "      <th>all uORFs</th>\n",
       "      <th>all ouORFs</th>\n",
       "      <th>all mean uORF lengths</th>\n",
       "      <th>all max uORF lengths</th>\n",
       "      <th>AUG_uORFs</th>\n",
       "      <th>CUG_uORFs</th>\n",
       "      <th>...</th>\n",
       "      <th>ACG_frame1_mean_uORF_length</th>\n",
       "      <th>ACG_frame1_max_uORF_length</th>\n",
       "      <th>ACG_frame2_uORFs</th>\n",
       "      <th>ACG_frame2_ouORFs</th>\n",
       "      <th>ACG_frame2_mean_uORF_length</th>\n",
       "      <th>ACG_frame2_max_uORF_length</th>\n",
       "      <th>ACG_frame3_uORFs</th>\n",
       "      <th>ACG_frame3_ouORFs</th>\n",
       "      <th>ACG_frame3_mean_uORF_length</th>\n",
       "      <th>ACG_frame3_max_uORF_length</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SeqID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000641515.2</th>\n",
       "      <td>CCCAGAUCUCUUCAGUUUUUAUGCCUCAUUCUGUGAAAAUUGCUGU...</td>\n",
       "      <td>Human</td>\n",
       "      <td>60</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.500000</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000616016.5</th>\n",
       "      <td>GGCGGCGGAGUCUCCCAAGUCCCCGCCGGGCGGGCGCGCGCCAGUG...</td>\n",
       "      <td>Human</td>\n",
       "      <td>509</td>\n",
       "      <td>0.809430</td>\n",
       "      <td>18.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>158.306667</td>\n",
       "      <td>318.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>...</td>\n",
       "      <td>18.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>297.0</td>\n",
       "      <td>297.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>228.0</td>\n",
       "      <td>228.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000437963.5</th>\n",
       "      <td>CAGCGCUUGGGGCUCGCGGGCCGCUCCCUCCGCUCGGAAGGGAAAA...</td>\n",
       "      <td>Human</td>\n",
       "      <td>60</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000342066.8</th>\n",
       "      <td>GCAGAGCCCAGCAGAUCCCUGCGGCGUUCGCGAGGGUGGGACGGGA...</td>\n",
       "      <td>Human</td>\n",
       "      <td>90</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>36.000000</td>\n",
       "      <td>42.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.0</td>\n",
       "      <td>42.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000338591.8</th>\n",
       "      <td>GGGAGUGAGCGACACAGAGCGGGCCGCCACCGCCGAGCAGCCCUCC...</td>\n",
       "      <td>Human</td>\n",
       "      <td>110</td>\n",
       "      <td>0.745455</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>63.000000</td>\n",
       "      <td>63.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_399996</th>\n",
       "      <td>UAGAAACGGCCGCAUCGACUCCGAGCCGACACGCCGUAGAGAGCGG...</td>\n",
       "      <td>Random_dl_gc_adjusted</td>\n",
       "      <td>236</td>\n",
       "      <td>0.546610</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>108.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>108.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>9.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_399997</th>\n",
       "      <td>CCAGUCCUGGAGGUGCUCAAGCGGCGGUGGAGCACAGCUCGUUCGG...</td>\n",
       "      <td>Random_dl_gc_adjusted</td>\n",
       "      <td>161</td>\n",
       "      <td>0.627329</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>33.500000</td>\n",
       "      <td>69.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_399998</th>\n",
       "      <td>CUGAGUGGUGUCAGGGCCGUACCCAAUCUACACCUAGUGACGGGGA...</td>\n",
       "      <td>Random_dl_gc_adjusted</td>\n",
       "      <td>113</td>\n",
       "      <td>0.592920</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>29.833333</td>\n",
       "      <td>51.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_399999</th>\n",
       "      <td>UCGGAAGAUACAAAUUGCCCUGCGGCCUAUAUGGAAAGACAGCCUU...</td>\n",
       "      <td>Random_dl_gc_adjusted</td>\n",
       "      <td>524</td>\n",
       "      <td>0.559160</td>\n",
       "      <td>48.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>51.641905</td>\n",
       "      <td>216.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>102.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>seq_400000</th>\n",
       "      <td>UCGCGCUACGUCGCGCCUACCGUCCGUCGGCGUGGGCUCUUCGGGG...</td>\n",
       "      <td>Random_dl_gc_adjusted</td>\n",
       "      <td>87</td>\n",
       "      <td>0.701149</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>12.000000</td>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482954 rows Ã— 88 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "seqs.to_csv('data/uORF_stats.csv')",
   "id": "9d62ae7ef552f06c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-05T09:58:37.882639Z",
     "start_time": "2024-09-05T09:58:37.767768Z"
    }
   },
   "cell_type": "code",
   "source": "humanseqs=seqs.loc[seqs['Group']=='Human']",
   "id": "6a787ded6dfb9963",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "start_time": "2024-09-16T08:24:36.768600Z"
    }
   },
   "cell_type": "code",
   "source": "cleanedhumanseqs=src.human_sequences.readFASTA(\"data/gencode_5utrs_human_cleaned.fa\")",
   "id": "35416bcc1218257f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "\n",
    "cleanedhumanseqs=src.human_sequences.delDupes(cleanedhumanseqs)\n",
    "#np.savetxt(\"data/humanseqsRNA.txt\",humanseqs,delimiter=\"\\t\",newline=\"\\n\",fmt='%s')"
   ],
   "id": "7639479d4bd8ae9"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "cleanedhumanseqs",
   "id": "35f873cdf324e8c6",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T09:00:02.939896Z",
     "start_time": "2024-09-16T09:00:02.609096Z"
    }
   },
   "cell_type": "code",
   "source": [
    "seqs = []\n",
    "seqIDs = []\n",
    "for seq_record in SeqIO.parse(\"data/gencode_5utrs_human_cleaned.fa\", \"fasta\"):\n",
    "    seqs.append(str(seq_record.seq))\n",
    "    seqIDs.append(seq_record.id)"
   ],
   "id": "a2a917062050b102",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T09:00:24.792510Z",
     "start_time": "2024-09-16T09:00:23.826629Z"
    }
   },
   "cell_type": "code",
   "source": "seqs = np.vstack((np.array(seqIDs), np.array(seqs))).T",
   "id": "bf735d9ade484e08",
   "outputs": [],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T09:00:48.639557Z",
     "start_time": "2024-09-16T09:00:48.629769Z"
    }
   },
   "cell_type": "code",
   "source": "humangroup = np.array([['Human'] * len(seqs)])",
   "id": "3ee6685882ff3229",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "seqs = np.concatenate((seqs, humangroup.T), axis=1)#!!!!",
   "id": "e3fd3628e273a376",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "\n",
    "for i in range(len(seqs)):\n",
    "    seqs[i, 1] = seqs[i, 1].replace('T', 'U')"
   ],
   "id": "e7a0ba031da1b80d"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "76236b22f7c1d248",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "27a8bf3832585b21",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "human160seqs=humanseqs[humanseqs[\"Sequence Length\"] >= 160]\n",
    "for i in range(len(human160seqs['Sequence'])):\n",
    "    seq = human160seqs.iloc[i, 0]\n",
    "    human160seqs.iloc[i, 0] = seq[len(seq) - 160: len(seq)]\n",
    "human160seqs.loc[:, 'Sequence Length'] = human160seqs['Sequence'].apply(len)"
   ],
   "id": "97c9f0967a97111b",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "secstructs=[]\n",
    "mfes=[]\n",
    "for seq in human160seqs[\"Sequence\"]:\n",
    "    (ss,mfe)=vrna.fold(seq)\n",
    "    secstructs.append(ss)\n",
    "    mfes.append(mfe)"
   ],
   "id": "cd1351e945e1ecd5",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "human160seqs.loc[:,'MFE']=mfes\n",
    "human160seqs.loc[:,'ss']=secstructs"
   ],
   "id": "191bf601d5d830ef",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "human160seqs.loc[:,['Sequence','Group','Sequence Length','MFE','ss']]",
   "id": "d9d6cc73f3513083",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "human160seqs.loc[:,['Sequence','Group','Sequence Length','MFE','ss']].to_csv('/fast/AG_Ohler/jdemoli/bachelorgit/data/MFE_human160_stats.csv')",
   "id": "843fe74e72e32cb0",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "1c9b2cea04ca2c25",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "4fa53f3b5b45d972",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "sns.histplot(human160seqs,x='MFE')",
   "id": "b1ae8ac69a1642ea",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "98ca2f2dcb49f573",
   "metadata": {},
   "source": [
    "'''Wir kreieren Nebendataframes um das Darstellen der Daten zu erleichtern'''\n",
    "gc50seqs=seqs.loc[seqs['Group']=='Random_gc50']\n",
    "gc70seqs=seqs.loc[seqs['Group']=='Random_gc70']\n",
    "lengthdistributed=pd.concat([humanseqs,gc50dlseqs])"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "humanseqs=seqs.loc[seqs['Group']=='Human']\n",
    "gc50dlseqs=seqs.loc[seqs['Group']=='Random_dl_gc_adjusted']"
   ],
   "id": "b018b731eb76a672",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7034beffe6cc5589",
   "metadata": {},
   "source": [
    "humanseqs160=humanseqs.loc[humanseqs[\"Sequence Length\"] <= 160]\n",
    "humanseqs160=humanseqs160[humanseqs160[\"Sequence Length\"]>=100]\n",
    "humanseqs100=humanseqs.loc[humanseqs[\"Sequence Length\"]<=100]\n",
    "print(len(humanseqs160)/len(humanseqs))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e3887879a0810911",
   "metadata": {},
   "source": [
    "#Sequenzlaenge\n",
    "sns.violinplot(data=seqs,x='Group',y='Sequence Length');"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4d06b4400af9e9ca",
   "metadata": {},
   "source": [
    "#GC Content \n",
    "%matplotlib \n",
    "sns.violinplot(data=seqs,x='Group',y='GC Content');\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b25c5de704707bee",
   "metadata": {},
   "source": [
    "# uORF Anzahl(Summe aus allen reading frames)\n",
    "sns.violinplot(data=seqs,x='Group',y='all uORFs');\n",
    "plt.ylim(0,100)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4bde7a2f002c2a06",
   "metadata": {},
   "source": [
    "#GC Content vs. uORF anzahl\n",
    "fig,axs = plt.subplots(nrows=1, ncols=2,layout='constrained' )\n",
    "im1= axs[0].hist2d(gc50seqs['GC Content'],gc50seqs['all uORFs'],norm=colors.LogNorm(),cmap='inferno',bins=10)\n",
    "im2= axs[1].hist2d(gc70seqs['GC Content'],gc70seqs['all uORFs'],norm=colors.LogNorm(),cmap='inferno',bins=8)\n",
    "axs[0].set_title(\"50% GC Content\")\n",
    "axs[1].set_title(\"70% GC Content\")\n",
    "#fig.colorbar(im1[3],ax=axs[0])\n",
    "#fig.colorbar(im2[3],ax=axs[1])\n",
    "fig.suptitle('GC Content vs uORF counts',fontsize=14)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8df5fea333c71b7",
   "metadata": {},
   "source": [
    "    #Menge an overlapping uORFs pro Gruppe\n",
    "im1=sns.histplot(x=seqs[\"AUG_ouORFs\"],hue=seqs[\"Group\"],multiple=\"dodge\",binwidth=0.25)\n",
    "im1.set_yscale('log')\n",
    "im1.set_xticks([0,1,2,3])\n",
    "im1.set_xlabel(\"ouORFs\")\n",
    "plt.title('ouORFs by Group')\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4f5e56fc980dbc7a",
   "metadata": {},
   "source": [
    "#overlapping uORFs in allen reading Frames\n",
    "fig2, axs2=plt.subplots(1,2,layout='constrained',sharey=True,sharex=True,figsize=(10,5) )\n",
    "sns.histplot(x=humanseqs[\"GUG_uORFs\"],multiple=\"dodge\",ax=axs2[0],bins=30)\n",
    "sns.histplot(x=humanseqs[\"GUG_frame3_uORFs\"],multiple=\"dodge\",ax=axs2[1],legend=False,bins=20,label='Frame 3')\n",
    "sns.histplot(x=humanseqs[\"GUG_frame2_uORFs\"],multiple=\"dodge\",ax=axs2[1],legend=False,bins=20,label='Frame 2')\n",
    "sns.histplot(x=humanseqs[\"GUG_frame1_uORFs\"],multiple=\"dodge\",ax=axs2[1],legend=False,bins=20,label='Frame 1')\n",
    "for ax in axs2:\n",
    "        ax.set_yscale('log')\n",
    "        ax.set_xlabel(\"\")\n",
    "axs2[0].set_title(\"Alle Frames\")\n",
    "axs2[1].set_title(\"Einzelne Frames\")\n",
    "\n",
    "plt.legend(loc='upper right')\n",
    "fig2.suptitle('uORFs in all reading frames',fontsize=14)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "79b539f8ac34308",
   "metadata": {},
   "source": [
    "- Frame 1 has more uORFs then 2 and 3 -> possible necessity/use of uORFs?\n",
    "- similar phenomenon is not observed with CUG uORFs\n",
    "- slightly more with GUG uORFs"
   ]
  },
  {
   "cell_type": "code",
   "id": "6d9d421fd65d60c1",
   "metadata": {},
   "source": [
    "fig3, axs3 =plt.subplots(2,2,layout='tight',sharey=True,figsize=(10,5.5))\n",
    "sns.histplot(x=gc50seqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[0,0])\n",
    "sns.histplot(x=gc70seqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[0,1])\n",
    "sns.histplot(x=humanseqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[1,0])\n",
    "sns.histplot(x=gc50dlseqs['AUG_max_uORF_length'],stat=\"probability\",ax=axs3[1,1])\n",
    "for arr in axs3:\n",
    "    for ax in arr:\n",
    "        ax.minorticks_on()\n",
    "        ax.set_xlabel(\"\")\n",
    "        #ax.set_xlim(0,100)\n",
    "        #ax.set_yscale('log')\n",
    "axs3[0,0].set_title(\"Random GC=50\")\n",
    "axs3[0,1].set_title(\"Random GC=70\")\n",
    "axs3[1,0].set_title(\"Human\")\n",
    "axs3[1,1].set_title(\"Random GC=50 - sequence length distributed\")\n",
    "fig3.suptitle(\"Length distribution of uORFs\",fontsize=14)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "a01b69c6974496f9",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "d78c9f7145d06d3d",
   "metadata": {},
   "source": [
    "- higher GC Content: More ouORFs, less full uORFs"
   ]
  },
  {
   "cell_type": "code",
   "id": "c4d32a4e6666c0ca",
   "metadata": {},
   "source": [
    "hue_order=['Random_dl_gc50','Human']\n",
    "sns.histplot(data=lengthdistributed,stat='probability', x='GUG_uORFs',common_norm=False, hue='Group',hue_order=hue_order)\n",
    "plt.xscale('log')\n",
    "#plt.yscale('log')\n",
    "#plt.xticks(np.arange(1,31));\n",
    "#plt.ylim(0.01,10000)\n",
    "plt.title(\"uORF count distribution\")\n",
    "plt.show()\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "cd0566d994d9676d",
   "metadata": {},
   "source": [
    "|- Human sequences are depleted of AUG uORFs(over 50% have none)\n",
    "- Less depleted in GUG uORFs(close to 40%)\n",
    "- even slgihtly less in CUG uORFs( a bit over 30%)\n",
    "- Random Sequences at 27%"
   ]
  },
  {
   "cell_type": "code",
   "id": "da2693b97f1f16de",
   "metadata": {},
   "source": [
    "hue_order=['Random_dl_gc50','Human']\n",
    "sns.kdeplot(data=lengthdistributed, x='Sequence Length', hue='Group',hue_order=hue_order,log_scale=True)\n",
    "plt.title(\"Sequence Length distribution\")\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "fca87535179f0ec3",
   "metadata": {},
   "source": [
    "fig4,axs4=plt.subplots(2,2,layout='constrained',sharex=True,sharey=True,figsize=(10,5))\n",
    "\n",
    "sns.histplot(data=lengthdistributed,x=\"all uORFs\",hue=\"Group\",stat='percent',ax=axs4[0,0],bins=200,hue_order=hue_order,common_norm=False)\n",
    "sns.histplot(lengthdistributed,x=\"AUG_uORFs\",hue=\"Group\",stat='percent',ax=axs4[0,1],bins=200,hue_order=hue_order,common_norm=False)\n",
    "sns.histplot(lengthdistributed,x=\"CUG_uORFs\",hue=\"Group\",stat='percent',ax=axs4[1,0],bins=200,hue_order=hue_order,common_norm=False)\n",
    "sns.histplot(lengthdistributed,x=\"GUG_uORFs\",hue=\"Group\",stat='percent',ax=axs4[1,1],bins=200,hue_order=hue_order,common_norm=False)\n",
    "for arr in axs4:\n",
    "    for ax in arr:\n",
    "        ax.set_xlim(0,100)\n",
    "    \n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "#MFE\n",
    "fig5, axs5 = plt.subplots(1,2,layout='constrained',sharex=True,sharey=True,figsize=(10,5))\n",
    "sns.scatterplot(humanseqs,x=\"Sequence Length\",y=\"MFE\",ax=axs5[0],label='Human')\n",
    "sns.scatterplot(gc50dlseqs,x=\"Sequence Length\",y=\"MFE\",ax=axs5[1],label='Random')\n",
    "plt.savefig('data/MFE.png')\n"
   ],
   "id": "16fd681ab0938bba",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "946b7ef55630ef0e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "fig6, axs6 = plt.subplots(1,2,layout='constrained',sharex=True,sharey=True,figsize=(10,5))\n",
    "sns.histplot(x=humanseqs[\"MFE\"]/humanseqs[\"Sequence Length\"],ax=axs6[0])\n",
    "sns.histplot(x=gc50dlseqs[\"MFE\"]/gc50dlseqs[\"Sequence Length\"],ax=axs6[1])\n",
    "for ax in axs6:\n",
    "    ax.set_xlabel('MFE per bp')\n",
    "plt.savefig('data/MFE_per_bp.png')"
   ],
   "id": "f099284b19011893",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "409475d2463028a3",
   "metadata": {},
   "source": "# Metadata"
  },
  {
   "cell_type": "code",
   "id": "bddc84b6c672c546",
   "metadata": {},
   "source": [
    "'''Plot of runtimes of the uORF_statistics method over the size of the data (number or length of sequences)'''\n",
    "times=[0.212,1.920,19.402,200.295]\n",
    "sizes=[100,1000,10000,100000]\n",
    "plt.plot(sizes,times)\n",
    "plt.xlabel(\"Size of the data\")\n",
    "plt.ylabel(\"Time [s]\")\n",
    "plt.title(\"runtime of oORFs algorithm vs size of data\")\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "acab16779abe6c80",
   "metadata": {},
   "source": [
    "times=[0.021,0.256,37.857,3753]\n",
    "sizes=[10,100,1000,10000]\n",
    "plt.plot(sizes,times)\n",
    "plt.xlabel(\"Size of the data\")\n",
    "plt.ylabel(\"Time [s]\")\n",
    "plt.title(\"runtime of outdated randomlengthseqs algorithm vs size of data\")\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "id": "82162840b11f58b2",
   "metadata": {},
   "source": [
    "# testing"
   ]
  },
  {
   "cell_type": "code",
   "id": "32ca32eb3abb208a",
   "metadata": {},
   "source": [
    "def testfunc(a,b):\n",
    "    print(a+b)\n",
    "def testingtestfunc(func,a,b):\n",
    "    func(a,b)\n",
    "    \n",
    "testingtestfunc(testfunc,1,2)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "7f022bac03543c9d",
   "metadata": {},
   "source": [
    "print(sum)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ccbd1e77a846ab32",
   "metadata": {},
   "source": [
    "rng = np.random.default_rng(seed=3)\n",
    "a=rng.choice([0,1,3],3)\n",
    "b=np.array([0,1,2,3])\n",
    "for i in b[1:]:\n",
    "    print(i)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "c1be1d26771292cb",
   "metadata": {},
   "source": [
    "def lognormalfunc(x,myu,sigma,a):\n",
    "    return a*np.exp(-1*(np.log(x)-myu)**2/(2*sigma)**2)/(x*sigma*np.sqrt(2*np.pi))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b6371a721e871c65",
   "metadata": {},
   "source": [
    "sns.scatterplot(x=x_plot,y=lognormalfunc(x_plot,myu,sigma,factor_a),size=1,marker=\".\");"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3622c3b50dd7cf8",
   "metadata": {},
   "source": [
    "x_plot=np.arange(1,8000)\n",
    "distribution=lognormalfunc(x_plot,myu,sigma,factor_a)\n",
    "distribution=distribution/sum(distribution)\n",
    "#print(sum(distribution))\n",
    "rng=np.random.default_rng()\n",
    "lengths=rng.choice(x_plot, 100000 , p=distribution)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "57529f7f0c5b16d4",
   "metadata": {},
   "source": [
    "im3=sns.histplot(lengths,kde=True)\n",
    "im3.lines[0].set_color('crimson')\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9242eb4f848c4b8d",
   "metadata": {},
   "source": [
    "im4=sns.scatterplot(x=x_plot,y=distribution,size=1,marker=\".\")\n",
    "#im4.lines[0].set_color('crimson')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "26826a8230cd62c2",
   "metadata": {},
   "source": [
    "lengthdata=list(humanseqs['Sequence Length'])\n",
    "lengthdata=Counter(lengthdata)\n",
    "x,y=[],[]\n",
    "for i,j in lengthdata.items():\n",
    "    x.append(i)\n",
    "    y.append(j)\n",
    "y=np.array(y)\n",
    "sns.scatterplot(x=x,y=y,size=1,marker=\".\");"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "45b5c03cb6609d49",
   "metadata": {},
   "source": [
    "params, cov_matrix= curve_fit(lognormalfunc,x,y)\n",
    "x_plot=np.arange(1,8000)\n",
    "plt.plot(x_plot,lognormalfunc(x_plot,params[0],params[1],params[2]))\n",
    "plt.plot(x,y,linestyle='None',marker=\".\")\n",
    "\n",
    "myu=params[0]\n",
    "sigma=params[1]\n",
    "factor_a=params[2]\n",
    "print(myu,sigma,factor_a)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "46f47d2d9f3cd293",
   "metadata": {},
   "source": [
    "def poissonfunc(x,lamb,a):\n",
    "    return a*poisson.pmf(x,lamb)\n",
    "params1, cov_matrix1 = curve_fit(poissonfunc,x,y)\n",
    "x_plot1=(np.arange(0,8000))\n",
    "plt.plot(x_plot1,poissonfunc(x_plot1,params1[0],params1[1]))\n",
    "plt.plot(x,y,linestyle='None',marker=\".\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "3370ab8c3217201c",
   "metadata": {},
   "source": [
    "x_plot=np.arange(1,8000)\n",
    "def softmax(x):\n",
    "    return np.exp(x)/sum(np.exp(x))\n",
    "\n",
    "    \n",
    "#plt.plot(x_plot,lognormalfunc(x_plot,myu,sigma))\n",
    "#plt.plot(x_plot,softmax(lognormalfunc(x_plot,myu,sigma)))\n",
    "#plt.plot(x_plot,scipy.special.softmax(lognormalfunc(x_plot,myu,sigma)))\n",
    "print([sum(lognormalfunc(x_plot,myu,sigma)),sum(softmax(lognormalfunc(x_plot,myu,sigma)))])"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "261e4e29a9dc30f0",
   "metadata": {},
   "source": [
    "#New approach for uORFs function\n",
    "def uORFs(seqs):\n",
    "    frames=src.uORF_Statistics.codons(seqs)\n",
    "    template={\n",
    "        \"uORFs\":list(np.zeros(len(frames))),\n",
    "        \"ouORFs\":list(np.zeros(len(frames))),\n",
    "        \"mean_uORF_length\":list(np.zeros(len(frames))),\n",
    "        \n",
    "        \"CUG_uORFs\":list(np.zeros(len(frames))),\n",
    "        \"CUG_ouORFs\":list(np.zeros(len(frames))),\n",
    "        \"mean_CUG_uORF_length\":list(np.zeros(len(frames))),\n",
    "        \n",
    "        \n",
    "        \"GUG_uORFs\":list(np.zeros(len(frames))),\n",
    "        \"GUG_ouORFs\":list(np.zeros(len(frames))),\n",
    "        \"mean_GUG_uORF_length\":list(np.zeros(len(frames))),\n",
    "    }\n",
    "    framedict={}\n",
    "    for frame in range(3):\n",
    "        framedict[f\"frame{frame+1}\"]=pd.DataFrame(template)\n",
    "        for index in range(len(frames[f\"frame{frame+1}\"])):\n",
    "            seq=frames[f\"frame{frame+1}\"][index]\n",
    "            lengths=[]\n",
    "            CUG_lengths=[]\n",
    "            GUG_lengths=[]\n",
    "            for codon in range(len(seq)):\n",
    "                if seq[codon]=='AUG':\n",
    "                    uORF_length=0\n",
    "                    ouORF_bool=True\n",
    "                    for nextcodon in range(len(seq)-codon-1):\n",
    "                        uORF_length+=3\n",
    "                        if seq[nextcodon+codon+1] in src.constants.stopcodons:\n",
    "                            ouORF_bool=False\n",
    "                            framedict[f\"frame{frame+1}\"].loc[index,\"uORFs\"]+=1\n",
    "                            lengths.append(uORF_length)\n",
    "                            break\n",
    "                    if ouORF_bool:\n",
    "                        framedict[f\"frame{frame+1}\"].loc[index,\"ouORFs\"]+=1\n",
    "                        break\n",
    "                    \n",
    "                if seq[codon]=='CUG':\n",
    "                    uORF_length=0\n",
    "                    ouORF_bool=True\n",
    "                    for nextcodon in range(len(seq)-codon-1):\n",
    "                        uORF_length+=3\n",
    "                        if seq[nextcodon+codon+1] in src.constants.stopcodons:\n",
    "                            ouORF_bool=False\n",
    "                            framedict[f\"frame{frame+1}\"].loc[index,\"CUG_uORFs\"]+=1\n",
    "                            CUG_lengths.append(uORF_length)\n",
    "                            break\n",
    "                    if ouORF_bool:\n",
    "                        framedict[f\"frame{frame+1}\"].loc[index,\"CUG_ouORFs\"]+=1\n",
    "                        break\n",
    "                        \n",
    "                if seq[codon]=='GUG':\n",
    "                    uORF_length=0\n",
    "                    ouORF_bool=True\n",
    "                    for nextcodon in range(len(seq)-codon-1):\n",
    "                        uORF_length+=3\n",
    "                        if seq[nextcodon+codon+1] in src.constants.stopcodons:\n",
    "                            ouORF_bool=False\n",
    "                            framedict[f\"frame{frame+1}\"].loc[index,\"GUG_uORFs\"]+=1\n",
    "                            GUG_lengths.append(uORF_length)\n",
    "                            break\n",
    "                    if ouORF_bool:\n",
    "                        framedict[f\"frame{frame+1}\"].loc[index,\"GUG_ouORFs\"]+=1\n",
    "                        break\n",
    "             \n",
    "            if len(lengths)>0:\n",
    "                framedict[f\"frame{frame+1}\"].loc[index,\"mean_uORF_length\"] = sum(lengths)/len(lengths)   \n",
    "            \n",
    "            if len(CUG_lengths)>0:\n",
    "                framedict[f\"frame{frame+1}\"].loc[index,\"mean_CUG_uORF_length\"]=sum(CUG_lengths)/len(CUG_lengths)\n",
    "            if len(GUG_lengths)>0:\n",
    "                framedict[f\"frame{frame+1}\"].loc[index,\"mean_GUG_uORF_length\"]=sum(GUG_lengths)/len(GUG_lengths)\n",
    "             \n",
    "    counts=pd.DataFrame()\n",
    "    \n",
    "    for key in framedict:\n",
    "        for column in framedict[key].columns:\n",
    "            counts[f\"{key}_{column}\"] = framedict[key].loc[:,column]\n",
    "            \n",
    "    counts[\"uORFs\"]=counts[\"frame1_uORFs\"]+counts[\"frame2_uORFs\"]+counts[\"frame3_uORFs\"]\n",
    "    counts[\"CUG_uORFs\"]=counts[\"frame1_CUG_uORFs\"]+counts[\"frame2_CUG_uORFs\"]+counts[\"frame3_CUG_uORFs\"]\n",
    "    counts[\"GUG_uORFs\"]=counts[\"frame1_GUG_uORFs\"]+counts[\"frame2_GUG_uORFs\"]+counts[\"frame3_GUG_uORFs\"]\n",
    "    counts[\"ouORFs\"]=counts[\"frame1_ouORFs\"]+counts[\"frame2_ouORFs\"]+counts[\"frame3_ouORFs\"]\n",
    "    counts[\"CUG_ouORFs\"]=counts[\"frame1_CUG_ouORFs\"]+counts[\"frame2_CUG_ouORFs\"]+counts[\"frame3_CUG_ouORFs\"]\n",
    "    counts[\"GUG_ouORFs\"]=counts[\"frame1_GUG_ouORFs\"]+counts[\"frame2_GUG_ouORFs\"]+counts[\"frame3_CUG_ouORFs\"]\n",
    "    counts[\"all uORFs\"]=counts[\"uORFs\"]+counts[\"CUG_uORFs\"]+counts[\"GUG_uORFs\"]\n",
    "    counts[\"all ouORFs\"]=counts[\"ouORFs\"]+counts[\"CUG_ouORFs\"]+counts[\"GUG_ouORFs\"]\n",
    "    \n",
    "    counts = counts.replace(0, np.nan)\n",
    "    counts[\"mean_uORF_length\"] = counts.loc[:,[\"frame1_mean_uORF_length\",\"frame2_mean_uORF_length\",\"frame3_mean_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"mean_CUG_uORF_length\"] = counts.loc[:,[\"frame1_mean_CUG_uORF_length\",\"frame2_mean_CUG_uORF_length\",\"frame3_mean_CUG_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"mean_GUG_uORF_length\"] = counts.loc[:,[\"frame1_mean_GUG_uORF_length\",\"frame2_mean_GUG_uORF_length\",\"frame3_mean_GUG_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"all mean lengths\"] = counts.loc[:,[\"mean_uORF_length\",\"mean_CUG_uORF_length\",\"mean_GUG_uORF_length\"]].mean(axis=1)\n",
    "    counts = counts.replace(np.nan, 0)\n",
    "\n",
    "    cols_in_front = [\"all uORFs\",\"all ouORFs\",\"all mean lengths\",\"uORFs\",\"CUG_uORFs\",\"GUG_uORFs\",\"ouORFs\",\"CUG_ouORFs\",\"GUG_ouORFs\",\"mean_uORF_length\",\"mean_CUG_uORF_length\",\"mean_GUG_uORF_length\"]\n",
    "    counts = counts.loc[:,[c for c in cols_in_front if c in counts.columns] + [c for c in counts if c not in cols_in_front]]\n",
    "    \n",
    "    if type(seqs) == pd.core.frame.DataFrame:\n",
    "        counts.insert(loc=0,column='SeqID',value=seqs.index)\n",
    "        counts.set_index('SeqID', inplace=True)\n",
    "    \n",
    "    return counts"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "31dc884642c12fff",
   "metadata": {},
   "source": [
    "singleseq=\"CACCCCGCUCGCGCCCAGCAGCGGACAUGCUCCACGAGCGCUGACAGCGUCGCCGUGCAACAAUCCUGGCGCCUGCCGCCGGCGAGCGGUCGACCGACCCGCACCGGUUCUCUAUGCCCCGCUACGGGCGGCCCGAGGCUUUACUACCCGGGCGCUUACC\"\n",
    "singleseq1=\"CCCGGCCGCGUCUGACUGCGCACCCCGCAAGUGUGUUGUCCCGCUCCCCAGCAGCCCCCCCCCCCCCCCGCUCGGCGCAGGGCCGAUCGCCACUUCCCUACCCGCGCGCCCACUUAUAGUCCCCGACUACCGUGCGCGGUUACUGCAGAAUGUCCGCCCCU\"\n",
    "#singleseqcodons=src.uORF_Statistics.codons(singleseq1)\n",
    "counts=src.uORF_Statistics.uORFs(singleseq1)\n",
    "counts.T\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e0dede270c603401",
   "metadata": {},
   "source": [
    "singleseqcodons.T.to_csv(\"data/testcodons.csv\",sep=\"\\t\",index=False)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "",
   "id": "504acc582dcf792a",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "af199c5a23750d7b",
   "metadata": {},
   "source": [
    "mydict={\"frame1\":[10,11,12],\"frame2\":[20,21,22],\"frame3\":[30,31,32]}\n",
    "num=69\n",
    "df=pd.DataFrame(mydict)\n",
    "df2=pd.DataFrame()\n",
    "df[\"mean\"]=df.loc[:,[\"frame2\",\"frame3\"]].mean(axis=1)\n",
    "df"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d69d02cebafda0c",
   "metadata": {},
   "source": [
    "mydict1={\"feature1\":[0],\"feature2\":[0],\"feature3\":[0]}\n",
    "template=pd.DataFrame(mydict1)\n",
    "l=[]\n",
    "for frame in range(3):\n",
    "    r1=np.random.randint(1,10)\n",
    "    r2=np.random.randint(1,10)\n",
    "    r3=np.random.randint(1,10)\n",
    "    df=pd.DataFrame(mydict1)\n",
    "    df.loc[0,\"feature1\"]=r1\n",
    "    df.loc[0,\"feature2\"]=r2\n",
    "    df.loc[0,\"feature3\"]=r3\n",
    "    l.append(df)\n",
    "print(l[0])\n",
    "print(l[1])\n",
    "print(l[2])"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "40c07afb654c4dce",
   "metadata": {},
   "source": [
    "var1=10\n",
    "var2=var1\n",
    "var1=11\n",
    "var2"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "38000d91f2241bc4",
   "metadata": {},
   "source": [
    "for i in range(10):\n",
    "    print(i)\n",
    "    if i ==5 :\n",
    "        print(\"test\")\n",
    "        for j in range(3):\n",
    "            print(\"test\", 2)\n",
    "            if j==1:\n",
    "                print(\"break\")\n",
    "                break"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "8781eebfed8cc188",
   "metadata": {},
   "source": [
    "singleseq2=\"AGCGCAUGUAUGUAUGCUGCGCGC\"\n",
    "count=0\n",
    "streaks=[]\n",
    "streak=0\n",
    "streakstart=False\n",
    "for i in range(0,len(singleseq2),2):\n",
    "    print(singleseq2[i:i+2])\n",
    "    if singleseq2[i:i+2]==\"GC\" or singleseq2[i:i+2]==\"CG\":\n",
    "        count+=1\n",
    "        if i>=2 and singleseq2[i-2:i]==singleseq2[i:i+2]:\n",
    "            streak+=2\n",
    "    else:\n",
    "        if streak > 0:\n",
    "            streaks.append(streak)\n",
    "            streak=0\n",
    "for i in range(1,len(singleseq2),2):\n",
    "    print(singleseq2[i:i+2])\n",
    "    if singleseq2[i:i+2]==\"GC\" or singleseq2[i:i+2]==\"CG\":\n",
    "        count+=1\n",
    "        if i>=2 and singleseq2[i-2:i]==singleseq2[i:i+2]:\n",
    "            streak+=2\n",
    "    else:\n",
    "        if streak > 0:\n",
    "            streaks.append(streak)\n",
    "            streak=0\n",
    "\n",
    "print(count,streaks)\n",
    "    \n",
    "        \n",
    "        "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4a06972378fa0ebb",
   "metadata": {},
   "source": [
    "def fast_uORfs(seqs,startcodon):\n",
    "    \n",
    "    seq_pos={\"frame1\":[],\"frame2\":[],\"frame3\":[]}\n",
    "    if type(seqs)== pd.DataFrame:\n",
    "        seqstrings=seqs[\"Sequence\"].values\n",
    "    if type(seqs)== str:\n",
    "        seqstrings=[seqs]\n",
    "    len_loop=len(seqstrings)\n",
    "    template = {\n",
    "        \"uORFs\": list(np.zeros(len_loop)),\n",
    "        \"ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"max_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \n",
    "        \"frame1_uORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame1_ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame1_mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"frame1_max_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \n",
    "        \"frame2_uORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame2_ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame2_mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"frame2_max_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \n",
    "        \"frame3_uORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame3_ouORFs\": list(np.zeros(len_loop)),\n",
    "        \"frame3_mean_uORF_length\": list(np.zeros(len_loop)),\n",
    "        \"frame3_max_uORF_length\": list(np.zeros(len_loop)),\n",
    "    }\n",
    "    counts=pd.DataFrame(template)\n",
    "    \n",
    "    posdict={\"frame1\":[[] for _ in range(len(seqstrings))],\"frame2\":[[] for _ in range(len(seqstrings))],\"frame3\":[[] for _ in range(len(seqstrings))]}\n",
    "    for index in range(len(seqstrings)):\n",
    "        seq=seqstrings[index]\n",
    "        for i in range(len(seq)):\n",
    "            rev_i=len(seq)-i\n",
    "            frame=(rev_i%3)\n",
    "            if seq[i:i+3]==startcodon:\n",
    "                posdict[f'frame{frame+1}'][index].append(i+1)\n",
    "            if seq[i:i+3]in src.constants.stopcodons:\n",
    "                posdict[f'frame{frame+1}'][index].append((i+1)*-1) \n",
    "\n",
    "    \n",
    "    for frame in range(3):\n",
    "        poslist=posdict[f'frame{frame+1}']\n",
    "        for index in range(len_loop):\n",
    "            pos=poslist[index]\n",
    "            if len(pos)>1:\n",
    "                lengths=[]\n",
    "                if pos[-1]>0:\n",
    "                    counts.loc[index,\"ouORFs\"]+=1\n",
    "                    counts.loc[index,f\"frame{frame+1}_ouORFs\"]+=1\n",
    "                    \n",
    "                for i in range(len(pos)):\n",
    "                    if pos[i]>0:\n",
    "                    \n",
    "                        for j in range(len(pos)-i-1):        \n",
    "                            if pos[i+j+1]<0:\n",
    "                                \n",
    "                                counts.loc[index,\"uORFs\"]+=1\n",
    "                                counts.loc[index,f\"frame{frame+1}_uORFs\"]+=1\n",
    "                                lengths.append((pos[i]+pos[i+j+1])*-1+3)\n",
    "                                break\n",
    "                if len(lengths)>0:\n",
    "                    counts.loc[index,f\"frame{frame+1}_mean_uORF_length\"]=sum(lengths)/len(lengths)\n",
    "                    counts.loc[index,f\"frame{frame+1}_max_uORF_length\"]=max(lengths)\n",
    "                if len(pos)==1 and pos[0]>0:\n",
    "                    counts.loc[index,\"ouORFs\"]+=1\n",
    "                    counts.loc[index,f\"frame{frame+1}_ouORFs\"]+=1\n",
    "                    \n",
    "                    \n",
    "                    \n",
    "    \n",
    "    counts[\"mean_uORF_length\"]=counts.loc[:,[\"frame1_mean_uORF_length\",\"frame2_mean_uORF_length\",\"frame3_mean_uORF_length\"]].mean(axis=1)\n",
    "    counts[\"max_uORF_length\"]=counts.loc[:,[\"frame1_max_uORF_length\",\"frame2_max_uORF_length\",\"frame3_max_uORF_length\"]].max(axis=1)\n",
    "    \n",
    "    return counts\n",
    "                 "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "49d120eb10d6e9e6",
   "metadata": {},
   "source": [
    "singleseq=\"AUGGCCGCAUAAAGAUGGAAUAGC\""
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "511771a86a5c848b",
   "metadata": {},
   "source": [
    "AUG_counts2=fast_uORfs(seqs,'AUG')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "18018709ad1544f",
   "metadata": {},
   "source": [
    "\n",
    "CUG_counts2=fast_uORfs(seqs,'CUG')\n",
    "GUG_counts2=fast_uORfs(seqs,'GUG')\n",
    "counts2=src.uORF_Statistics.counts_concat_all(seqs,AUG_counts2,CUG_counts2,GUG_counts2)\n",
    "counts2"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "84878300d2bc1816",
   "metadata": {},
   "source": [
    "counts1"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d1afc5b5002239f8",
   "metadata": {},
   "source": [
    "seqstrings[-1]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "c82a75caa76612eb",
   "metadata": {},
   "source": [
    "l=\"AUGUGAUAAAUGCAGAU\"\n",
    "frame1=[]\n",
    "frame2=[]\n",
    "frame3=[]\n",
    "for i in range(len(l)-2):\n",
    "    rev_i=len(l)-i\n",
    "    if  rev_i%3==0:\n",
    "        frame1.append(l[i]+l[i+1]+l[i+2])\n",
    "    if rev_i %3==1:\n",
    "        frame2.append(l[i]+l[i+1]+l[i+2])\n",
    "    if rev_i %3==2:\n",
    "        frame3.append(l[i]+l[i+1]+l[i+2])\n",
    "print(frame1,'\\n',frame2,'\\n',frame3)        "
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "cf2935a60f014141",
   "metadata": {},
   "source": [
    "l=[3,21,24,2,1,2,0,-1,4]\n",
    "print(max(l))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5590152ebd727324",
   "metadata": {},
   "source": [
    "l300=np.arange(400)\n",
    "for i in range(4000):\n",
    "    for i in range(len(l300)):\n",
    "        for j in range(len(l300)-1):\n",
    "            a=i+j\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "human160seqs = humanseqs[humanseqs[\"Sequence Length\"] >= 160]\n",
    "human160seqs"
   ],
   "id": "6b7ab7057b6e2f45",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "human160seqs['Sequence_adjusted']=humanseqs['Sequence']",
   "id": "a6d75e82233967fa",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T08:23:35.749564Z",
     "start_time": "2024-09-16T08:23:35.745925Z"
    }
   },
   "cell_type": "code",
   "source": "print(src.constants.DNABASES)",
   "id": "af326f9a44650d10",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A', 'C', 'G', 'T']\n"
     ]
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2024-09-16T08:20:26.712024Z",
     "start_time": "2024-09-16T08:20:26.710079Z"
    }
   },
   "cell_type": "code",
   "source": "",
   "id": "fa08b9423be686f1",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "f26dfbbb2ded4470"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
